{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coverage and Quality Plots\n",
    "\n",
    "Input: \n",
    "1. Gene counts table from FeatureCounts\n",
    "2. Gene Body Coverage Files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## LOAD LIBRARIES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.default_inference import DefaultInference\n",
    "from pydeseq2.ds import DeseqStats\n",
    "import re\n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import gtfparse as gtf_parse\n",
    "import seaborn as sns\n",
    "import os.path\n",
    "import itertools\n",
    "import matplotlib.gridspec as gridspec\n",
    "import copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_gtf(gtf_path):\n",
    "    genes = []\n",
    "    opener = gzip.open if gtf_path.endswith(\".gz\") else open\n",
    "    with opener(gtf_path, \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\") or \"\\tgene\\t\" not in line:\n",
    "                continue\n",
    "            parts = line.strip().split(\"\\t\")\n",
    "            chrom, source, feature, start, end, score, strand, frame, attributes = parts\n",
    "            gene_id = None\n",
    "            gene_name = None\n",
    "            gene_type = None\n",
    "            for attr in attributes.split(\";\"):\n",
    "                if \"gene_id\" in attr:\n",
    "                    gene_id = attr.split('\"')[1]\n",
    "                if \"gene_name\" in attr:\n",
    "                    gene_name = attr.split('\"')[1]\n",
    "                if \"gene_type\" in attr:\n",
    "                    gene_type = attr.split('\"')[1]\n",
    "                if gene_id and gene_name and gene_type:\n",
    "                    break\n",
    "            if gene_id:\n",
    "                genes.append({\n",
    "                    \"gene_id\": gene_id,\n",
    "                    \"chrom\": chrom,\n",
    "                    \"start\": int(start),\n",
    "                    \"end\": int(end),\n",
    "                    \"strand\": strand,\n",
    "                    \"gene_name\" : gene_name,\n",
    "                    \"gene_type\" : gene_type\n",
    "                })\n",
    "    return genes\n",
    "\n",
    "GTF_PATH = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/BEDMETHYL_files/last_filter_round/test/gencode.v43.sorted.gtf\"  # Can be gzipped (.gtf.gz)\n",
    "gene_list = parse_gtf(GTF_PATH)\n",
    "genes_df = pd.DataFrame.from_dict(gene_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. TRANSCRIPTOME-WIDE GENE COUNTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Read featureCounts table from all ONT samples\n",
    "counts_file=\"RNA002_RNA004.DORADO.0.7.2.GRCh38.featurecounts.tsv\"\n",
    "counts_file=\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/DORADO_0_7_2_folder/RNA002_RNA004.DORADO.0.7.2.GRCh38.featurecounts.tsv\"\n",
    "counts = pd.read_csv(counts_file, sep = \"\\t\", header = 0, comment='#')\n",
    "counts = counts.set_index(\"Geneid\")\n",
    "counts = counts.drop(['Chr', 'Start', 'End', 'Strand', 'Length'], axis=1)\n",
    "counts = counts.drop(columns=[col for col in counts.columns if '_E' in col])\n",
    "counts = counts.drop(columns=[col for col in counts.columns if '_ribo' in col])\n",
    "print(counts.columns)\n",
    "counts.columns =['RNA002_HEK293T_A', 'RNA002_HEK293T_B', 'RNA002_blood_1',\n",
    "       'RNA002_blood_IVT_1', 'RNA002_UHRR', 'RNA004_HEK293T_A',\n",
    "       'RNA004_HEK293T_B', 'RNA004_blood_1', 'RNA004_blood_IVT_1', 'RNA004_UHRR_1',\n",
    "       'RNA004_UHRR_2', 'RNA004_UHRR_3']\n",
    "\n",
    "counts_file2=\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/S5_analysis_5_12_24/S5_RNA004.DORADO.0.7.2.GRCh38.featurecounts.tsv\"\n",
    "counts2 = pd.read_csv(counts_file2, sep = \"\\t\", header = 0, comment='#')\n",
    "counts2 = counts2.set_index(\"Geneid\")\n",
    "counts2 = counts2.drop(['Chr', 'Start', 'End', 'Strand', 'Length'], axis=1)\n",
    "print(counts2.columns)\n",
    "counts2.columns = [\"RNA004_blood_2\"]\n",
    "\n",
    "counts_file3=\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/S6_IVT_analysis/S6_RNA004.DORADO.0.7.2.GRCh38.featurecounts.tsv\"\n",
    "counts3 = pd.read_csv(counts_file3, sep = \"\\t\", header = 0, comment='#')\n",
    "counts3 = counts3.set_index(\"Geneid\")\n",
    "counts3 = counts3.drop(['Chr', 'Start', 'End', 'Strand', 'Length'], axis=1)\n",
    "print(counts3.columns)\n",
    "counts3.columns = [\"RNA004_blood_3\", \"RNA004_blood_IVT_3_R1\", \"RNA004_blood_IVT_3_R2\"]\n",
    "\n",
    "counts = pd.concat([counts, counts2, counts3], axis=1)\n",
    "\n",
    "\n",
    "metadata = pd.DataFrame({\"sample\" : counts.columns})\n",
    "metadata[\"tissue\"] = metadata[\"sample\"].str.replace(\"_[A|B|1|2|3].*$\", \"\", regex=True)\n",
    "metadata[\"chem_type\"] = metadata[\"tissue\"].str.replace(\"_.*$\", \"\", regex=True)\n",
    "metadata[\"ID\"] = metadata[\"sample\"].str.replace(\"RNA00[2|4]_\", \"\", regex=True)\n",
    "metadata[\"tissue_flat\"] = metadata[\"tissue\"].str.replace(\"RNA00[2|4]_\", \"\", regex = True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_tissue = counts.loc[:, ~counts.columns.str.contains('IVT')]\n",
    "metadata_tissue = metadata.loc[~metadata[\"sample\"].str.contains(\"IVT\"), :]\n",
    "\n",
    "# Extract genes covered by at least 10 reads in each sample type\n",
    "df_covered_genes = (counts_tissue >= 10).sum(axis=0)\n",
    "\n",
    "gene_counts_10 = pd.DataFrame(df_covered_genes)\n",
    "gene_counts_10 = gene_counts_10.reset_index()\n",
    "gene_counts_10.columns = [\"sample_type\", \"# genes covered >= 10\"]\n",
    "gene_counts_10  = pd.merge(metadata_tissue.reset_index(), gene_counts_10, left_on='sample', right_on='sample_type', how = \"left\")\n",
    "gene_counts_10_sub = gene_counts_10[[\"ID\", \"sample_type\", \"tissue_flat\", \"chem_type\", \"# genes covered >= 10\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_counts_10_sub.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_TABLES_genes_cov_10.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "counts_df = counts.reset_index()\n",
    "counts_df_annot = pd.merge(genes_df, counts_df.rename(columns ={\"Geneid\" : \"gene_id\"}), on = \"gene_id\", how = \"left\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FIGURE 3B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# --- Setup ---\n",
    "# Assuming `gene_counts_10_sub` is your DataFrame\n",
    "tissue_groups = gene_counts_10_sub.groupby(\"tissue_flat\")\n",
    "colors_input = {\"RNA002\": \"#ee6c4d\", \"RNA004\": \"#0b3954\", \"NGS\": \"#8b5cf6\"}\n",
    "hatches = [['','//', '', '//'],['', '', '//', '-'], ['', '', '//', '-'],['', '', '//', '//']]\n",
    "\n",
    "# Bar width settings\n",
    "bar_width = 0.6\n",
    "spacing = 0\n",
    "facet_widths = []\n",
    "\n",
    "# Calculate subplot widths based on number of bars in each facet\n",
    "facet_data = []\n",
    "for tissue, df in tissue_groups:\n",
    "    n_bars = df['sample_type'].nunique()\n",
    "    print(n_bars)\n",
    "    width = max(1, n_bars * (bar_width + spacing))\n",
    "    facet_data.append((tissue, df))\n",
    "    facet_widths.append(width)\n",
    "\n",
    "# Normalize widths for matplotlib gridspec\n",
    "total_width = sum(facet_widths)\n",
    "print(total_width)\n",
    "normalized_widths = [w / total_width for w in facet_widths]\n",
    "print(normalized_widths)\n",
    "# --- Create Subplots with variable widths ---\n",
    "fig_height = 3\n",
    "n_facets = len(facet_data)\n",
    "fig, axes = plt.subplots(1, n_facets, figsize=(5, fig_height), \n",
    "                         gridspec_kw={'width_ratios': normalized_widths})\n",
    "\n",
    "# Handle single facet case\n",
    "if n_facets == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "# --- Plotting each facet ---\n",
    "for c, (ax, (tissue, df)) in enumerate(zip(axes, facet_data)):\n",
    "    #print(df)\n",
    "    print(df.sort_values(\"ID\")[\"sample_type\"].unique())\n",
    "    order_list = df.sort_values(\"ID\")[\"sample_type\"].unique()\n",
    "   \n",
    "    sns.barplot(\n",
    "        data=df, y=\"# genes covered >= 10\", x=\"sample_type\", hue=\"chem_type\",\n",
    "        ax=ax, palette=colors_input,order  = order_list, edgecolor=\"black\", width=0.8\n",
    "    )\n",
    "    if c >= 1:\n",
    "        ax.tick_params(left=False, bottom=True, labelleft=False, labelbottom=True)\n",
    "    # Clean x labels\n",
    "    ax.set_xticklabels([t.get_text().replace(\"RNA002_\", \"\").replace(\"RNA004_\", \"\") \n",
    "                        for t in ax.get_xticklabels()], rotation=45, ha = \"right\", rotation_mode='anchor')\n",
    "    ax.grid(False)\n",
    "    \n",
    "\n",
    "    ax.set_title(tissue)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(\"# genes covered >= 10\" if c == 0 else \"\")\n",
    "\n",
    "    ax.legend([],[], frameon=False) \n",
    "\n",
    "    # Set hatches\n",
    "    for i, bar in enumerate(ax.patches):\n",
    "        if c < len(hatches) and i < len(hatches[c]):\n",
    "            bar.set_hatch(hatches[c][i])\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "    if tissue == \"Summed\":\n",
    "        \n",
    "        ax.spines[\"left\"].set_visible(True)\n",
    "        ax.tick_params(left=True, bottom=True, labelleft=True, labelbottom=True)\n",
    "   \n",
    "    if c >= 1 and tissue != \"Summed\":\n",
    "        ax.tick_params(left=False, bottom=True, labelleft=False, labelbottom=True)\n",
    "\n",
    "        ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.subplots_adjust(wspace=0.2, hspace=0)\n",
    "\n",
    "plt.show()\n",
    "\n",
    "\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW2.svg\", dpi = 300)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW2.pdf\", dpi = 300)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW2.png\", dpi = 300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. MENDELIOME-WIDE GENE COUNTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### LOAD MENDELIOM GENES \n",
    "#mendeliom = pd.read_csv(\"mendeliom_IHDB.tsv\", sep = \"\\t\")\n",
    "mendeliom = pd.read_csv(\"/raid/awiercze_analysis/RESEARCH/REVISION_FIGURES/mendeliom_IHDB.tsv\", sep = \"\\t\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "counts_ensembl =  counts_tissue\n",
    "counts_ensembl.index = [i.split('.')[0] for i in counts_ensembl.index]\n",
    "counts_ensembl[\"gene_id\"] = counts_ensembl.index\n",
    "counts_mendel = counts_ensembl[counts_ensembl[\"gene_id\"].isin(mendeliom.ensembl_gene_id)]\n",
    "counts_mendel = counts_mendel.drop(\"gene_id\", axis = 1)\n",
    "\n",
    "global_counts = counts_ensembl\n",
    "global_counts.index = [i.split('.')[0] for i in counts_ensembl.index]\n",
    "global_counts_mendel = global_counts[global_counts.index.isin(mendeliom.ensembl_gene_id)]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "global_counts_mendel_sub = global_counts_mendel.drop(\"gene_id\", axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLOT NUMBER OF DISTINCT MENDELIOME GENES WITH AT LEAST 10 READ COUNTS\n",
    "df_covered_genes = (global_counts_mendel_sub >= 10).sum(axis=0)\n",
    "gene_counts_10 = pd.DataFrame(df_covered_genes)\n",
    "gene_counts_10 = gene_counts_10.reset_index()\n",
    "gene_counts_10.columns = [\"sample_type\", \"# genes covered >= 10\"]\n",
    "gene_counts_10  = pd.merge(metadata.reset_index(), gene_counts_10, left_on='sample', right_on='sample_type', how = \"left\")\n",
    "gene_counts_10[\"tissue_flat\"] = gene_counts_10[\"tissue\"].str.replace(\"RNA00[2|4]_\", \"\", regex = True)\n",
    "gene_counts_10_sub2 = gene_counts_10.loc[~gene_counts_10[\"sample\"].str.contains(\"IVT\"),:]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plot2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "tissue_groups = gene_counts_10_sub2.groupby(\"tissue_flat\")\n",
    "colors_input = {\"RNA002\": \"#ee6c4d\", \"RNA004\": \"#0b3954\", \"NGS\": \"#8b5cf6\"}\n",
    "hatches = [['','//', '', '//'],['', '', '//', '-'], ['', '', '//', '-'],['', '', '//', '//']]\n",
    "\n",
    "\n",
    "bar_width = 0.6\n",
    "spacing = 0\n",
    "facet_widths = []\n",
    "\n",
    "facet_data = []\n",
    "for tissue, df in tissue_groups:\n",
    "    n_bars = df['sample_type'].nunique()\n",
    "    print(n_bars)\n",
    "    width = max(1, n_bars * (bar_width + spacing))\n",
    "    facet_data.append((tissue, df))\n",
    "    facet_widths.append(width)\n",
    "\n",
    "total_width = sum(facet_widths)\n",
    "print(total_width)\n",
    "normalized_widths = [w / total_width for w in facet_widths]\n",
    "\n",
    "fig_height = 3\n",
    "n_facets = len(facet_data)\n",
    "fig, axes = plt.subplots(1, n_facets, figsize=(5, fig_height), \n",
    "                        gridspec_kw={'width_ratios': normalized_widths})\n",
    "\n",
    "if n_facets == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "for c, (ax, (tissue, df)) in enumerate(zip(axes, facet_data)):\n",
    "\n",
    "    print(df.sort_values(\"ID\")[\"sample_type\"].unique())\n",
    "    order_list = df.sort_values(\"ID\")[\"sample_type\"].unique()\n",
    "    sns.barplot(\n",
    "        data=df, y=\"# genes covered >= 10\", x=\"sample_type\", hue=\"chem_type\",\n",
    "        ax=ax, palette=colors_input,order  = order_list, edgecolor=\"black\", width=0.8\n",
    "    )\n",
    "    if c >= 1:\n",
    "        ax.tick_params(left=False, bottom=True, labelleft=False, labelbottom=True)\n",
    "    ax.set_xticklabels([t.get_text().replace(\"RNA002_\", \"\").replace(\"RNA004_\", \"\") \n",
    "                        for t in ax.get_xticklabels()], rotation=45, ha = \"right\", rotation_mode='anchor')\n",
    "    ax.grid(False)\n",
    "    \n",
    "\n",
    "    ax.set_title(tissue)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(\"# Mendelian genes\" if c == 0 else \"\")\n",
    "\n",
    "    ax.legend([],[], frameon=False) \n",
    "    for i, bar in enumerate(ax.patches):\n",
    "        if c < len(hatches) and i < len(hatches[c]):\n",
    "            bar.set_hatch(hatches[c][i])\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "    if c >= 1 and tissue != \"Summed\":\n",
    "        ax.tick_params(left=False, bottom=True, labelleft=False, labelbottom=True)\n",
    "\n",
    "        ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.subplots_adjust(wspace=0.2, hspace=0)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW_MENDELIOM.svg\", dpi = 300)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW_MENDELIOM.pdf\", dpi = 300)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW_MENDELIOM.png\", dpi = 300)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Percentage of Mendeliome genes with coverage >= 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "\n",
    "# --- Setup ---\n",
    "# Assume gene_counts_10 is a DataFrame and each row corresponds to a sample\n",
    "# You must have a \"total_genes\" column or replace 20000 with actual total\n",
    "TOTAL_GENES = len(counts_mendel)  # replace if dynamic\n",
    "\n",
    "# Calculate percentage covered\n",
    "gene_counts_10 = gene_counts_10_sub2.copy()\n",
    "gene_counts_10[\"% genes covered >= 10\"] = (gene_counts_10[\"# genes covered >= 10\"] / TOTAL_GENES) * 100\n",
    "\n",
    "# Group by tissue\n",
    "tissue_groups = gene_counts_10.groupby(\"tissue_flat\")\n",
    "colors_input = {\"RNA002\": \"#ee6c4d\", \"RNA004\": \"#0b3954\", \"NGS\": \"#8b5cf6\"}\n",
    "hatches = [['','//', '', '//'],['', '', '//', '-'], ['', '', '//', '-'],['', '', '//', '//']]\n",
    "\n",
    "# Bar width settings\n",
    "bar_width = 0.6\n",
    "spacing = 0\n",
    "facet_widths = []\n",
    "\n",
    "facet_data = []\n",
    "for tissue, df in tissue_groups:\n",
    "    n_bars = df['sample_type'].nunique()\n",
    "    width = max(1, n_bars * (bar_width + spacing))\n",
    "    facet_data.append((tissue, df))\n",
    "    facet_widths.append(width)\n",
    "\n",
    "# Normalize subplot widths\n",
    "total_width = sum(facet_widths)\n",
    "normalized_widths = [w / total_width for w in facet_widths]\n",
    "\n",
    "# Create subplots\n",
    "fig_height = 3\n",
    "n_facets = len(facet_data)\n",
    "fig, axes = plt.subplots(1, n_facets, figsize=(5, fig_height), \n",
    "                         gridspec_kw={'width_ratios': normalized_widths})\n",
    "\n",
    "# Handle single subplot\n",
    "if n_facets == 1:\n",
    "    axes = [axes]\n",
    "\n",
    "# --- Plot each tissue facet ---\n",
    "for c, (ax, (tissue, df)) in enumerate(zip(axes, facet_data)):\n",
    "    order_list = df.sort_values(\"ID\")[\"sample_type\"].unique()\n",
    "   \n",
    "    sns.barplot(\n",
    "        data=df, y=\"% genes covered >= 10\", x=\"sample_type\", hue=\"chem_type\",\n",
    "        ax=ax, palette=colors_input, order=order_list, edgecolor=\"black\", width=0.8\n",
    "    )\n",
    "\n",
    "    if c >= 1:\n",
    "        ax.tick_params(left=False, bottom=True, labelleft=False, labelbottom=True)\n",
    "\n",
    "    ax.set_xticklabels(\n",
    "        [t.get_text().replace(\"RNA002_\", \"\").replace(\"RNA004_\", \"\") \n",
    "         for t in ax.get_xticklabels()],\n",
    "        rotation=45, ha=\"right\", rotation_mode='anchor'\n",
    "    )\n",
    "    \n",
    "    ax.grid(False)\n",
    "    ax.set_title(tissue)\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(\"% Mendelian genes\\nwith >= 10 reads\" if c == 0 else \"\")\n",
    "    ax.legend([], [], frameon=False)\n",
    "\n",
    "    # Hatches\n",
    "    for i, bar in enumerate(ax.patches):\n",
    "        if c < len(hatches) and i < len(hatches[c]):\n",
    "            bar.set_hatch(hatches[c][i])\n",
    "\n",
    "    ax.spines[\"top\"].set_visible(False)\n",
    "    ax.spines[\"right\"].set_visible(False)\n",
    "\n",
    "    if c >= 1 and tissue != \"Summed\":\n",
    "        ax.tick_params(left=False, bottom=True, labelleft=False, labelbottom=True)\n",
    "        ax.spines[\"left\"].set_visible(False)\n",
    "\n",
    "fig.tight_layout()\n",
    "fig.subplots_adjust(wspace=0.2, hspace=0)\n",
    "\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW_MENDELIOM_perc.svg\", dpi = 300)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW_MENDELIOM_perc.pdf\", dpi = 300)\n",
    "fig.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/PAPER_FIGURE_3A_NEW_MENDELIOM_perc.png\", dpi = 300)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SAVE NUMBER OF DISTINCT COVERED GENES\n",
    "gene_counts_10.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_TABLE_Mendeliom_genes_covered.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save regions of mendeliom genes as bed file for subsequent analyses\n",
    "# Extract mendeliome genes from sample-wise and technique-wise counts tables\n",
    "counts_file=\"RNA002_RNA004.DORADO.0.7.2.GRCh38.featurecounts.tsv\"\n",
    "counts = pd.read_csv(counts_file, sep = \"\\t\", header = 0, comment='#')\n",
    "counts = counts.set_index(\"Geneid\")\n",
    "counts_regions = counts[['Chr', 'Start', 'End', 'Strand', 'Length']]\n",
    "counts_regions.index = [i.split('.')[0] for i in counts_regions.index]\n",
    "counts_regions[\"gene_id\"] = counts_regions.index\n",
    "counts_mendel_tmp = counts_regions[counts_regions[\"gene_id\"].isin(mendeliom.ensembl_gene_id)]\n",
    "counts_mendel_tmp['Chr'] = counts_mendel_tmp['Chr'].str.split(';').str[0]\n",
    "counts_mendel_tmp['Start'] = counts_mendel_tmp['Start'].str.split(';').str[0]\n",
    "counts_mendel_tmp['End'] = counts_mendel_tmp['End'].str.split(';').str[-1]\n",
    "counts_mendel_tmp['Strand'] = counts_mendel_tmp['Strand'].str.split(';').str[0]\n",
    "counts_mendel_tmp[[\"Chr\", \"Start\", \"End\", \"Strand\", \"gene_id\"]].to_csv(\"mendeliome_gene_regions.bed\", index = False, header = False, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SAVE COUNTS TABLE INCLUDING MENDELIOME ANNOTATION PER GENE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df_annot[\"transcript\"] = [i.split('.')[0] for i in counts_df_annot[\"gene_id\"]]\n",
    "counts_df_annot[\"MENDELIOM\"] = [\"YES\" if i in global_counts_mendel.index else \"NO\" for i in counts_df_annot[\"transcript\"] ]\n",
    "counts_df_annot.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_TABLES_RAW_COUNTS_ANNOT.tsv\")\n",
    "counts_df_annot.to_excel(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_TABLES_RAW_COUNTS_ANNOT.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. DIFFERENTIAL EXPRESSION ANALYIS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_tissue.to_csv(\"/home/awiercze/metadata.tsv\", sep = \"\\t\", index = False)\n",
    "counts.to_csv(\"/home/awiercze/counts.tsv\", sep = \"\\t\", index = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# R CODE LOCAL: \n",
    "\"\"\"\"\n",
    "setwd(\"C:/Users/wie5a/Documents/RNA_MODIFCATIONS_DIAGNOSTICS\")\n",
    "\n",
    "# Load libraries\n",
    "library(DESeq2)\n",
    "library(ggplot2)\n",
    "library(pheatmap)\n",
    "\n",
    "\n",
    "# Load the count data\n",
    "counts_data <- read.table(\"counts.tsv\", sep = \"\\t\", header = TRUE, row.names = 1)\n",
    "# Load the sample information\n",
    "metadata <- read.table(\"metadata.tsv\", sep = \"\\t\", header = TRUE, row.names = 1, comment.char = \"\")\n",
    "\n",
    "metadata$tissue_flat = gsub(\"RNA00[2|4]_\", \"\", metadata$tissue)\n",
    "# Create a DESeqDataSet object\n",
    "dds <- DESeqDataSetFromMatrix(countData = counts_data, colData = metadata, design = ~1)\n",
    "# Pre-filtering: remove rows with low counts\n",
    "dds <- dds[rowSums(counts(dds)) > 10, ]\n",
    "# Run the DESeq2 pipeline\n",
    "dds <- DESeq(dds)\n",
    "\n",
    "# Plot PCA\n",
    "rld <- rlog(dds)\n",
    "pcaData <- plotPCA(rld, intgroup = c(\"chem_type\", \"tissue_flat\"), returnData = TRUE)\n",
    "percentVar <- round(100 * attr(pcaData, \"percentVar\"))  \n",
    "p <- ggplot(pcaData, aes(x = PC1, y = PC2, color = chem_type, shape = tissue_flat)) +\n",
    "  geom_point(size = 3) +\n",
    "  xlab(paste0(\"PC1: \", percentVar[1], \"% variance\")) +\n",
    "  ylab(paste0(\"PC2: \", percentVar[2], \"% variance\")) +\n",
    "  theme_minimal()\n",
    "# Save the PCA plot\n",
    "ggsave(\"RNA004_PCA_plot.png\", plot = p, width = 8, height = 6)\n",
    "\n",
    "# Save the PCA data for further analysis\n",
    "write.table(pcaData, file = \"PCA_data.tsv\", sep = \"\\t\", quote = FALSE, row.names = TRUE, col.names = TRUE)\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca_data = pd.read_csv(\"/home/awiercze/PCA_data.tsv\", sep = \"\\t\", header = 0, index_col = 0)\n",
    "percentVar = [80, 18] # FROM R SCRIPT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors_input = {\"RNA002\": \"#ee6c4d\", \"RNA004\": \"#0b3954\", \"NGS\" : \"#8b5cf6\"}\n",
    "plt.figure(figsize=(5, 3.4))  # Optional: Increase figure size for better resolution\n",
    "sns.scatterplot(\n",
    "    data=pca_data,\n",
    "    x=\"PC1\",\n",
    "    y=\"PC2\",\n",
    "    hue=\"chem_type\",\n",
    "    style=\"tissue_flat\",\n",
    "    palette=colors_input, markers={\"HEK293T\": \"o\", \"blood\": \"s\", \"blood_IVT\" : \"D\", \"UHRR\" : \"X\"},\n",
    "    s=150  # Increase dot size\n",
    ")\n",
    "\n",
    "# Change axis labels\n",
    "plt.xlabel(\"PC1 (\" + str(percentVar[0]) + \"% variance)\", fontsize=14)\n",
    "plt.ylabel(\"PC2 (\"+ str(percentVar[1]) + \"% variance)\", fontsize=14)\n",
    "plt.legend(bbox_to_anchor=(1.01, 0.99), loc='upper left', borderaxespad=0.)\n",
    "# Optional: Improve layout\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save as high-resolution image\n",
    "plt.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/pca_scatter_high_res.png\", dpi=300)  # Save at 300 DPI\n",
    "plt.savefig(\"/raid/awiercze_analysis/RNA004_REVISION_NAR/pca_scatter_high_res.pdf\", dpi=300)  # Save at 300 DPI\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4. GENE BODY COVERAGE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata.loc[metadata[\"sample\"].str.contains(\"_[2|3|B]\"), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "geneBod = pd.read_csv(\"/raid/awiercze_analysis/GeneBodyCov_RNA002_RNA004_0.7.2.tsv\", sep = \"\\t\", header = 0)\n",
    "metadata[\"Replicate\"] = 1\n",
    "metadata.loc[metadata[\"sample\"].str.contains(\"_[2|B]\"), \"Replicate\"] = 2\n",
    "metadata.loc[metadata[\"sample\"].str.contains(\"_3$\"), \"Replicate\"] = 3\n",
    "\n",
    "metadata.loc[metadata[\"sample\"].str.contains(\"_3_R1\"), \"Replicate\"] = 2\n",
    "metadata.loc[metadata[\"sample\"].str.contains(\"_3_R2\"), \"Replicate\"] = 3\n",
    "\n",
    "geneBod = geneBod.merge(metadata, left_on = \"ID\", right_on = \"sample\", how = \"left\")\n",
    "geneBod[\"Chemistry\"] = geneBod[\"Condition\"].str.replace(\"_.*\",\"\", regex = True)\n",
    "geneBod[\"tissue\"] = geneBod[\"tissue\"].str.replace(\"RNA00[2|4]_\", \"\", regex = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(7, 4))\n",
    "\n",
    "# Custom tissue color palette\n",
    "tissue_palette = {\n",
    "    \"HEK293T\": \"#0b3954\",\n",
    "    \"blood\": \"#098c9a\",\n",
    "    \"UHRR\": \"#ee6c4d\",\n",
    "    \"blood_IVT\": \"#8b5cf6\"\n",
    "}\n",
    "\n",
    "# Optional: define marker shapes for each unique 'Shape'\n",
    "linestyle_map = {\n",
    "    1: \"-\",\n",
    "    2: \"--\",\n",
    "    3: \"-.\",\n",
    "    4: \":\"\n",
    "}\n",
    "\n",
    "shape_marker_map = {\n",
    "    \"RNA002\": \"o\",\n",
    "    \"RNA004\": \"\"\n",
    "}\n",
    "\n",
    "# Plot each ID_x group separately\n",
    "for id_name, df_sub in geneBod.groupby(\"ID_x\"):\n",
    "    tissue = df_sub[\"tissue\"].iloc[0]\n",
    "    shape = df_sub[\"Replicate\"].iloc[0]\n",
    "    chemistry = df_sub[\"Chemistry\"].iloc[0]\n",
    "\n",
    "    linestyle = linestyle_map.get(shape, \"-\")  # default linestyle if missing\n",
    "    color = tissue_palette.get(tissue, \"gray\")  # default color if tissue missing\n",
    "    marker = shape_marker_map.get(chemistry, \"\")  # default marker if shape missing\n",
    "\n",
    "    sns.lineplot(\n",
    "        data=df_sub,\n",
    "        x=\"Percentiles\", y=\"PercVal\",\n",
    "        color=color,\n",
    "        marker=marker,\n",
    "        linestyle=linestyle,\n",
    "        markersize=4,\n",
    "        #markeredgecolor=\"black\",\n",
    "        #markeredgewidth=0.5,   \n",
    "        ax=ax,\n",
    "        label=id_name  # now works fine\n",
    "    )\n",
    "    for line in ax.lines:\n",
    "        line.set_markevery(5)\n",
    "\n",
    "# Format axes\n",
    "ax.set(xlabel='Gene Body Percentiles', ylabel='Relative Coverage')\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.grid(False)\n",
    "\n",
    "# Clean up legend\n",
    "sns.move_legend(ax, \"upper left\", bbox_to_anchor=(1, 1), title=\"Samples\")\n",
    "fig.tight_layout()\n",
    "\n",
    "fig.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE2/PAPER_geneBodyCoverage_tissue.png\", dpi = 300, transparent=True)\n",
    "fig.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE2/PAPER_geneBodyCoverage_tissue.pdf\", dpi = 300, transparent=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
