{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# -----PSEU PANEL 5------"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob as glob\n",
    "import pandas as pd\n",
    "from upsetplot import generate_counts\n",
    "from upsetplot import from_contents\n",
    "from upsetplot import UpSet\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pyranges as pr\n",
    "from itertools import combinations\n",
    "from sklearn.metrics import r2_score\n",
    "from scipy.stats import pearsonr\n",
    "from scipy import stats\n",
    "from matplotlib_venn import venn2, venn2_circles\n",
    "from matplotlib_venn import venn3, venn3_circles\n",
    "\n",
    "# Set seaborn plotting theme\n",
    "custom_params = {\"axes.spines.right\": False, \"axes.spines.top\": False}\n",
    "sns.set_theme(style=\"ticks\", rc=custom_params)\n",
    "\n",
    "colors_input = {\n",
    "  \"RNA004_HEK293T_A\" : \"#EE6C4D\",\n",
    "  \"RNA004_HEK293T_B\" : \"#E69581\",\n",
    "  \"RNA004_blood_1\" :\"#0B3954\",\n",
    "  \"RNA004_blood_IVT_1\" :\"#8b5cf6\",\n",
    "  \"RNA004_UHRR_1\" : \"#098C9A\",\n",
    "  \"RNA004_UHRR_2\" : \"#26AFBE\",\n",
    "  \"RNA004_UHRR_3\" : \"#93C6CC\", \n",
    "  \"RNA004_blood_2\" : \"#177BB5\",\n",
    "  \"RNA004_blood_3\" : \"#38A6E5\",\n",
    "  \"RNA004_blood_IVT_3\" : \"#BA9EFA\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LOAD PSEU SITES FROM MODKIT"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## BLOOD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blood_RNA004_pseu = glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/m6A_pseu_refilter_rev1_nar1/*blood*pse*\") + glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/m6A_pseu_refilter_rev1_nar1/*S*DRS*pse*\")\n",
    "new_blood_RNA004_pseu = \"/home/awiercze/RNA004_S5_DRS_basecall.0.7.2.GRCh38_pse*.r1.mod.rna002_cov.bed\"\n",
    "new_blood3_RNA004_pseu = \"/home/awiercze/RNA004_S6_DRS_basecall.0.7.2.GRCh38_pse*.r1.mod.rna002_cov.bed\"\n",
    "blood_RNA004_pseu.append(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/S6_IVT_analysis/merged_RNA004_S6_IVT_pseu.r1.mod.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_pseu = pd.DataFrame()\n",
    "    \n",
    "for j in blood_RNA004_pseu:\n",
    "    print(j)\n",
    "    t_tmp = pd.read_csv(j, sep = \"\\t\", header = None)\n",
    "    print(len(t_tmp))\n",
    "\n",
    "    t_tmp.columns = ['chrom', 'chromStart', 'chromEnd', 'name', 'score', 'strand', \"mod_start\", \"mod_end\", \"color\", \"coverage\", \"modRatio\", \"n_mod\", \"n_canonical\", \"n_other_mod\", \"n_delete\", \"n_fail\", \"n_diff\", \"n_nocall\"]\n",
    "    caller = \"Dorado@v5.0.0\"\n",
    "    t_tmp = t_tmp[[\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\", \"coverage\", \"modRatio\"]]\n",
    "    t_tmp[\"sample\"] = j.split(\"/\")[-1].replace(\"_basecall.0.7.2.GRCh38_pseu.r1.mod.bed\", \"\").replace(\".0.7.2.GRCh38_pseu.r1.mod.bed\", \"\").replace(\"merged_RNA004_S6_IVT_pseu.r1.mod.bed\", \"RNA004_S6_IVT_merged\")\n",
    "\n",
    "    t_tmp[\"caller\"] = caller\n",
    "    t_tmp[\"chemistry\"] = \"RNA004\"\n",
    "    p_pseu = pd.concat([p_pseu, t_tmp], axis = 0)\n",
    "p_pseu = p_pseu.reset_index()\n",
    "p_pseu = p_pseu.drop(\"index\", axis = 1)\n",
    "p_pseu[\"chromStart\"] = p_pseu[\"chromStart\"].astype(int)\n",
    "p_pseu[\"chromEnd\"] = p_pseu[\"chromEnd\"].astype(int)\n",
    "p_pseu[\"coverage\"] = p_pseu[\"coverage\"].astype(int)\n",
    "p_pseu[\"modRatio\"] = p_pseu[\"modRatio\"].astype(float)\n",
    "    \n",
    "# Create unique identifier per position\n",
    "p_pseu[\"pos\"] = p_pseu[\"chrom\"] + \":\" + p_pseu[\"chromStart\"].astype(str) + \"\" + p_pseu[\"strand\"]\n",
    "\n",
    "# Use only sites on chr1-22,Y,X,M\n",
    "p_pseu = p_pseu[p_pseu[\"chrom\"].str.contains(\"chr\")]\n",
    "# Filter table for covered sites \n",
    "p_pseu_cov10 = p_pseu[(p_pseu[\"coverage\"] >= 10)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## CELL-LINES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_lines_RNA004_pseu = glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/m6A_pseu_refilter_rev1_nar1/*_A*pse*.bed\") + glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/m6A_pseu_refilter_rev1_nar1/*_B*pse*\") + glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/m6A_pseu_refilter_rev1_nar1/*UHRR*pse*\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_cell_pseu = pd.DataFrame()\n",
    "    \n",
    "for j in cell_lines_RNA004_pseu:\n",
    "    print(j)\n",
    "    t_tmp = pd.read_csv(j, sep = \"\\t\", header = None)\n",
    "    print(len(t_tmp))\n",
    "\n",
    "    t_tmp.columns = ['chrom', 'chromStart', 'chromEnd', 'name', 'score', 'strand', \"mod_start\", \"mod_end\", \"color\", \"coverage\", \"modRatio\", \"n_mod\", \"n_canonical\", \"n_other_mod\", \"n_delete\", \"n_fail\", \"n_diff\", \"n_nocall\"]\n",
    "    caller = \"Dorado@v5.0.0\"\n",
    "    t_tmp = t_tmp[[\"chrom\", \"chromStart\", \"chromEnd\", \"name\", \"score\", \"strand\", \"coverage\", \"modRatio\"]]\n",
    "    t_tmp[\"sample\"] = j.split(\"/\")[-1].replace(\"_basecall.0.7.2.GRCh38_pseu.r1.mod.bed\", \"\").replace(\".0.7.2.GRCh38_pseu.r1.mod.bed\", \"\").replace(\"merged_RNA004_S6_IVT_pseu.r1.mod.bed\", \"RNA004_S6_IVT_merged\")\n",
    "\n",
    "    t_tmp[\"caller\"] = caller\n",
    "    t_tmp[\"chemistry\"] = \"RNA004\"\n",
    "    p_cell_pseu = pd.concat([p_cell_pseu, t_tmp], axis = 0)\n",
    "p_cell_pseu = p_cell_pseu.reset_index()\n",
    "p_cell_pseu = p_cell_pseu.drop(\"index\", axis = 1)\n",
    "p_cell_pseu[\"chromStart\"] = p_cell_pseu[\"chromStart\"].astype(int)\n",
    "p_cell_pseu[\"chromEnd\"] = p_cell_pseu[\"chromEnd\"].astype(int)\n",
    "p_cell_pseu[\"coverage\"] = p_cell_pseu[\"coverage\"].astype(int)\n",
    "p_cell_pseu[\"modRatio\"] = p_cell_pseu[\"modRatio\"].astype(float)\n",
    "    \n",
    "# Create unique identifier per position\n",
    "p_cell_pseu[\"pos\"] = p_cell_pseu[\"chrom\"] + \":\" + p_cell_pseu[\"chromStart\"].astype(str) + \"\" + p_cell_pseu[\"strand\"]\n",
    "\n",
    "\n",
    "p_cell_pseu = p_cell_pseu[p_cell_pseu[\"chrom\"].str.contains(\"chr\")]\n",
    "# Filter table for covered sites\n",
    "p_cell_pseu_cov10 = p_cell_pseu[(p_cell_pseu[\"coverage\"] >= 10)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Concatenate blood, UHRR and HEK293T pseU site tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Concatenate Blood and Cell-lines table of all pseu sites\n",
    "pseu_all_samples = pd.concat([p_pseu, p_cell_pseu], axis = 0)\n",
    "replace_dict = {\n",
    "'RNA004_blood' : 'RNA004_blood_1',\n",
    " 'RNA004_blood_IVT' : 'RNA004_blood_IVT_1',\n",
    " 'RNA004_S6_IVT_merged' : 'RNA004_blood_IVT_3',\n",
    " 'RNA004_S5_DRS' : 'RNA004_blood_2',\n",
    " 'RNA004_S6_DRS': 'RNA004_blood_3',\n",
    " 'RNA004_A' : 'RNA004_HEK293T_A',\n",
    " 'RNA004_B': 'RNA004_HEK293T_B',\n",
    "}\n",
    "pseu_all_samples.loc[:, \"sample\"] = pseu_all_samples[\"sample\"].replace(replace_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CREATE MAIN OUTPUT TABLE PSEU"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD GENE ANNOTATION FROM GENCODE\n",
    "def parse_gtf(gtf_path):\n",
    "    genes = []\n",
    "    opener = gzip.open if gtf_path.endswith(\".gz\") else open\n",
    "    with opener(gtf_path, \"rt\") as f:\n",
    "        for line in f:\n",
    "            if line.startswith(\"#\") or \"\\tgene\\t\" not in line:\n",
    "                continue\n",
    "            parts = line.strip().split(\"\\t\")\n",
    "            chrom, source, feature, start, end, score, strand, frame, attributes = parts\n",
    "            gene_id = None\n",
    "            gene_name = None\n",
    "            gene_type = None\n",
    "            for attr in attributes.split(\";\"):\n",
    "                if \"gene_id\" in attr:\n",
    "                    gene_id = attr.split('\"')[1]\n",
    "                if \"gene_name\" in attr:\n",
    "                    gene_name = attr.split('\"')[1]\n",
    "                if \"gene_type\" in attr:\n",
    "                    gene_type = attr.split('\"')[1]\n",
    "                if gene_id and gene_name and gene_type:\n",
    "                    break\n",
    "            if gene_id:\n",
    "                genes.append({\n",
    "                    \"gene_id\": gene_id,\n",
    "                    \"chrom\": chrom,\n",
    "                    \"start\": int(start),\n",
    "                    \"end\": int(end),\n",
    "                    \"strand\": strand,\n",
    "                    \"gene_name\" : gene_name,\n",
    "                    \"gene_type\" : gene_type\n",
    "                })\n",
    "    return genes\n",
    "\n",
    "GTF_PATH = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/BEDMETHYL_files/last_filter_round/test/gencode.v43.sorted.gtf\"  # Can be gzipped (.gtf.gz)\n",
    "gene_list = parse_gtf(GTF_PATH)\n",
    "genes_df = pd.DataFrame.from_dict(gene_list)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rmbase_sites[\"pos\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# LOAD STOPP CODON SITES FROM GENCODE\n",
    "stopp_codons = pd.read_csv(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/TEST_folder/ANNO_misc/gencodev43.stopp_codon.bed\", sep  = \"\\t\", header = None, names = [\"Chromosome\", \"Start\", \"End\", \"type\", \"Strand\", \"Transcript\"])\n",
    "stopp_codons[\"type\"] = \"STOP CODON\"\n",
    "\n",
    "# LOAD pseU SITES FROM RMBase\n",
    "cols_mod_file = [\"Chromosome\",\"Start\",\"End\",\"modID\", \"Score\", \"Strand\",\"modType\", \"supportNum\", \"supportList\", \"supportListSub\",\"pubList\",\"cellList\",\"seqTypeList\",\"geneID\", \"transcriptID\",\"geneName\",\"geneType\" ,\"Region\", \"Seq\", \"motifScore\",\"conservedSitesList\",\"snoRNA_detailInfo\",\"snoRNA_guideSite\",\"snoRNA_nameList\",\"snoRNA_dataBase\",\"writerLoc\",\"writerID\",\"writerNameList\",\"source\"]\n",
    "rmbase_sites = pd.read_csv(\"~/NEW_RNA004_PLOTS/human.hg38.Pseudo.result.col29.bed\", sep = \"\\t\", header = None, names = cols_mod_file)\n",
    "rmbase_sites = rmbase_sites[rmbase_sites[\"Chromosome\"].str.contains(\"chr\")]\n",
    "rmbase_sites[\"type\"] = \"RMBase\"\n",
    "rmbase_sites[\"pos\"] =  rmbase_sites[\"Chromosome\"].astype(str) + \":\" + rmbase_sites[\"Start\"].astype(str) + rmbase_sites[\"Strand\"].astype(str)\n",
    "\n",
    "# LOAD pseU SITES FROM TAVAKOLI et al.\n",
    "tava = \"/home/jensenone/psu_high_conf_0base.bed\"\n",
    "tava_sites = pd.read_csv(tava, sep = \"\\t\", header = None)\n",
    "tava_sites[\"type\"] = \"Tavakoli et al.\"\n",
    "tava_sites.columns = [\"Chromosome\", \"Start\", \"End\", \"type\"]\n",
    "tava_sites[\"pos\"] = tava_sites[\"Chromosome\"].astype(str) + \":\" + tava_sites[\"Start\"].astype(str) \n",
    "\n",
    "# LOAD pseU SITES FROM PRAISE\n",
    "praise = \"/raid/awiercze_analysis/RNA004_REVISION_NAR/PRAISE_pseU_sites.txt\"\n",
    "praise_sites = pd.read_csv(praise, sep = \"\\t\")\n",
    "praise_sites = praise_sites[~praise_sites[\"chr_site\"].str.contains(\"NONE\")]\n",
    "praise_sites[\"Chromosome\"] = [i.split(\"_\")[0] for i in praise_sites[\"chr_site\"]]\n",
    "praise_sites = praise_sites[praise_sites[\"Chromosome\"].str.contains(\"chr\")]\n",
    "praise_sites[\"Position\"] = [i.split(\"_\")[1] for i in praise_sites[\"chr_site\"]]\n",
    "praise_sites = praise_sites[~praise_sites[\"Position\"].str.contains(\"K\")]\n",
    "praise_sites = praise_sites[~praise_sites[\"Position\"].str.contains(\"G\")]\n",
    "praise_sites = praise_sites[~praise_sites[\"Position\"].str.contains(\"M\")]\n",
    "praise_sites[\"Start\"] = [int(i.split(\"-\")[0]) if \"-\" in i else i for i in praise_sites[\"Position\"] ]\n",
    "praise_sites[\"End\"] = [i.split(\"-\")[1] if \"-\" in i else int(i) + 1 for i in praise_sites[\"Position\"] ]\n",
    "praise_sites = praise_sites[[\"Chromosome\", \"Start\", \"End\", \"gene_name\"]]\n",
    "praise_sites[\"type\"] = \"PRAISE\"\n",
    "strands = []\n",
    "new_starts = []\n",
    "new_ends = []\n",
    "for i, row in praise_sites.iterrows():\n",
    "    start = int(row[\"Start\"])\n",
    "    end = int(row[\"End\"])\n",
    "    if start > end: \n",
    "        strands.append(\"-\")\n",
    "        new_starts.append(end)\n",
    "        new_ends.append(start)\n",
    "    else:\n",
    "        strands.append(\"+\")\n",
    "        new_starts.append(start)\n",
    "        new_ends.append(end)\n",
    "praise_sites[\"Start\"] = new_starts\n",
    "praise_sites[\"End\"] = new_ends\n",
    "praise_sites[\"Strand\"] = strands"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ANNOTATE ALL SITES "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Add tissue label per site \n",
    "pseu_all_samples[\"sample_type\"] = \"blood\"\n",
    "pseu_all_samples.loc[pseu_all_samples[\"sample\"].str.contains(\"UHRR\"), \"sample_type\"] = \"UHRR\"\n",
    "pseu_all_samples.loc[pseu_all_samples[\"sample\"].str.contains(\"HEK293T\"), \"sample_type\"] = \"HEK293T\"\n",
    "pseu_all_samples.loc[pseu_all_samples[\"sample\"].str.contains(\"IVT\"), \"sample_type\"] = \"blood_IVT\"\n",
    "\n",
    "# Filter table to sites with min. 10 read coverage \n",
    "pseu_all_samples_cov = pseu_all_samples[pseu_all_samples[\"coverage\"] >= 10]\n",
    "# Filter table to sites with min. 10 read coverage and modRatio >= 10\n",
    "pseu_all_samples_cov_pseU = pseu_all_samples[(pseu_all_samples[\"coverage\"] >= 10) & (pseu_all_samples[\"modRatio\"] >= 10)]\n",
    "\n",
    "# Create copy of table\n",
    "df = pseu_all_samples[pseu_all_samples.pos.isin(pseu_all_samples_cov_pseU[\"pos\"].drop_duplicates())]\n",
    "\n",
    "\n",
    "pos_cols = ['chrom', 'chromStart', 'chromEnd', 'strand']\n",
    "\n",
    "# Check if site is covered in at least sample per tissue\n",
    "df['cov>=10.RNA004_Dorado_blood'] = (df['coverage'] >= 10) & (df['sample_type'] == 'blood')\n",
    "df['cov>=10.RNA004_Dorado_blood_IVT'] = (df['coverage'] >= 10) & (df['sample_type'] == 'blood_IVT')\n",
    "df['cov>=10.RNA004_Dorado_UHRR'] = (df['coverage'] >= 10) & (df['sample_type'] == 'UHRR')\n",
    "df['cov>=10.RNA004_Dorado_HEK293T'] = (df['coverage'] >= 10) & (df['sample_type'] == 'HEK293T')\n",
    "\n",
    "# Check if site is covered and modified in at least sample per tissue\n",
    "df['pseU_ratio>=10.RNA004_Dorado_blood'] =(df['coverage'] >= 10) & (df['modRatio'] >= 10) & (df['sample_type'] == 'blood')\n",
    "df['pseU_ratio>=10.RNA004_Dorado_blood_IVT'] =(df['coverage'] >= 10) & (df['modRatio'] >= 10) & (df['sample_type'] == 'blood_IVT')\n",
    "df['pseU_ratio>=10.RNA004_Dorado_UHRR'] = (df['coverage'] >= 10) & (df['modRatio'] >= 10) & (df['sample_type'] == 'UHRR')\n",
    "df['pseU_ratio>=10.RNA004_Dorado_HEK293T'] =(df['coverage'] >= 10) & (df['modRatio'] >= 10) & (df['sample_type'] == 'HEK293T')\n",
    "\n",
    "df_grouped = df.groupby(pos_cols).agg({\n",
    "    'cov>=10.RNA004_Dorado_blood': 'any',\n",
    "    'cov>=10.RNA004_Dorado_blood_IVT': 'any',\n",
    "    'cov>=10.RNA004_Dorado_UHRR': 'any',\n",
    "    'cov>=10.RNA004_Dorado_HEK293T': 'any',\n",
    "    'pseU_ratio>=10.RNA004_Dorado_blood': 'any',\n",
    "    'pseU_ratio>=10.RNA004_Dorado_blood_IVT': 'any',\n",
    "    'pseU_ratio>=10.RNA004_Dorado_UHRR': 'any',\n",
    "    'pseU_ratio>=10.RNA004_Dorado_HEK293T': 'any'}).astype(int).reset_index()\n",
    "\n",
    "# Extract modRatio per site and sample\n",
    "modratios = df.pivot_table(index=pos_cols, columns='sample', values='modRatio', aggfunc='first').reset_index()\n",
    "modratios.columns.name = None  # remove index name\n",
    "samples = [\"pseU_freq.\" + s for s in modratios.columns[4:].to_list()]\n",
    "my_dict = dict(zip(modratios.columns[4:].to_list(), samples))\n",
    "\n",
    "# Extract mean modRatio per tissue and site\n",
    "mean_modratios_cells = df.groupby(pos_cols + ['sample_type'])['modRatio'].mean().unstack().reset_index()\n",
    "mean_modratios_cells.columns.name = None\n",
    "mean_modRatio_Dorado_cells = mean_modratios_cells.rename(columns={\n",
    "    \"blood\" : 'avg_pseU_ratio.RNA004_Dorado_blood',\n",
    "    \"blood_IVT\" : 'avg_pseU_ratio.RNA004_Dorado_blood_IVT',\n",
    "    'UHRR': 'avg_pseU_ratio.RNA004_Dorado_UHRR',\n",
    "    'HEK293T': 'avg_pseU_ratio.RNA004_Dorado_HEK293T',\n",
    "})\n",
    "\n",
    "# Extract coverage per site and sample\n",
    "coverages = df.pivot_table(index=pos_cols, columns='sample', values='coverage', aggfunc='first').reset_index()\n",
    "samples = [\"n_reads.\" + s for s in coverages.columns[4:].to_list()]\n",
    "my_dict = dict(zip(coverages.columns[4:].to_list(), samples))\n",
    "coverages = coverages.rename(columns = my_dict)\n",
    "\n",
    "# Calculate number of tissues with modified site\n",
    "df['modified'] = ((df['coverage'] >= 10) & (df['modRatio'] >= 10)).astype(int)\n",
    "detected_tissues = df[(df['modified'] == 1)].groupby(pos_cols)['sample_type'].nunique().reset_index(name='Dorado_tissues_detected')\n",
    "\n",
    "# Merge all subtables\n",
    "result = df_grouped.merge(mean_modRatio_Dorado_cells, on=pos_cols, how='outer').merge(detected_tissues, on = pos_cols, how = \"outer\").merge(coverages, on=pos_cols, how = 'outer').merge(modratios, on=pos_cols, how='outer')\n",
    "\n",
    "# Remove sites that are not modified in any sample\n",
    "result_sub = result[~result[\"Dorado_tissues_detected\"].isna()]\n",
    "\n",
    "# Add gene annotation per site\n",
    "query_ranges = pr.PyRanges(result_sub.rename(columns = {\"chrom\": \"Chromosome\", \"chromStart\" : \"Start\", \"chromEnd\" : \"End\", \"strand\" : \"Strand\"}))\n",
    "gtf_ranges = pr.PyRanges(genes_df.rename(columns = {\"chrom\": \"Chromosome\", \"start\" : \"Start\", \"end\" : \"End\", \"strand\" : \"Strand\"}))\n",
    "result_annot = gtf_ranges.join(query_ranges, how=\"right\", suffix = \"_Pos\")\n",
    "result_annot_df = result_annot.df.rename(columns = {\"Start\" : \"Start_gene\", \"End\" : \"End_gene\", \"Start_Pos\" : \"Start\", \"End_Pos\" : \"End\"})\n",
    "# Add pseu Databases annotation per site\n",
    "query_stopp = (pr.PyRanges(result_annot_df)\n",
    "    .join(pr.PyRanges(stopp_codons), how = \"left\", suffix = \"_SC\")\n",
    "    .join(pr.PyRanges(rmbase_sites), suffix = \"_RMBase\", how = \"left\")\n",
    "    .join(pr.PyRanges(praise_sites), how=\"left\", suffix = \"_PRAISE\")\n",
    "    .join(pr.PyRanges(tava_sites), how=\"left\", suffix = \"_Tavakoli\")\n",
    ")\n",
    "result_annot2_df = query_stopp.df\n",
    "\n",
    "# Reformat table\n",
    "result_annot_df_sub = result_annot2_df[result_annot2_df.columns[[5,0,6, 1,8,4,]].to_list() + result_annot2_df.columns[[45,76, 80, 84]].to_list() + result_annot2_df.columns[10:43].to_list()] \n",
    "result_annot_df_sub = result_annot_df_sub.rename(columns = {\"type\" : \"stop codon\", \"type_RMBase\" : \"RMBase\", \"type_PRAISE\" : \"PRAISE\", \"type_Tavakoli\" : \"Tavakoli et al.\"})\n",
    "# Save table\n",
    "result_annot_df_sub.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/PAPER_TABLE_POSITIONS_pseU_annot_NEW.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CREATE ALL CORRELATIONS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "result_rows = []\n",
    "df = pseu_all_samples\n",
    "\n",
    "# Extract unique samples\n",
    "samples = df['sample'].unique()\n",
    "\n",
    "# Iterate over all pairwise combinations of all samples to calculate pseu ratio correlations between shared pseU sites\n",
    "for s1, s2 in combinations(samples, 2):\n",
    "    print(s1 + \" vs \" + s2)\n",
    "\n",
    "        \n",
    "    df1 = df[df['sample'] == s1]\n",
    "    df2 = df[df['sample'] == s2]\n",
    "\n",
    "    # Filtered sets\n",
    "    df1_cov = df1[df1['coverage'] >= 10]\n",
    "    df2_cov = df2[df2['coverage'] >= 10]\n",
    "\n",
    "    df1_mod = df1_cov[df1_cov['modRatio'] >= 10]\n",
    "    df2_mod = df2_cov[df2_cov['modRatio'] >= 10]\n",
    "\n",
    "    \n",
    "    # All positions (just with coverage >= 10)\n",
    "    pos_cov1 = set(df1_cov['pos'])\n",
    "    pos_cov2 = set(df2_cov['pos'])\n",
    "\n",
    "    # Positions with coverage >= 10 and modRatio >= 10\n",
    "    pos_mod1 = set(df1_mod['pos'])\n",
    "    pos_mod2 = set(df2_mod['pos'])\n",
    "\n",
    "    # Joined positions\n",
    "    joined_cov = pos_cov1 & pos_cov2\n",
    "    joined_cov_mod = pos_mod1 & pos_mod2\n",
    "    \n",
    "    # Overlap input:\n",
    "    df1_mod_joined = df1_mod[df1_mod.pos.isin(joined_cov)]\n",
    "    df2_mod_joined = df2_mod[df2_mod.pos.isin(joined_cov)]\n",
    "    \n",
    "    if joined_cov_mod:\n",
    "        df1_r2 = df1_mod[df1_mod['pos'].isin(joined_cov_mod)].set_index('pos')\n",
    "        df2_r2 = df2_mod[df2_mod['pos'].isin(joined_cov_mod)].set_index('pos')\n",
    "        mod1 = df1_r2.loc[df1_r2.index.isin(joined_cov_mod), 'modRatio']\n",
    "        mod2 = df2_r2.loc[df1_r2.index.isin(joined_cov_mod), 'modRatio']\n",
    "        r = pearsonr(mod1, mod2)\n",
    "        r = r[0]\n",
    "        corr = round(r, 4)\n",
    "        \n",
    "        #input_set1 = df1_r2\n",
    "        #input_set2 = df2_r2\n",
    "        #m = pd.merge(input_set1.reset_index(), input_set2.reset_index(), how = \"inner\", on = \"pos\")\n",
    "        #g = sns.jointplot(x=\"modRatio_x\", y=\"modRatio_y\", data=m, kind='hex', color=\"#0B3954\",     height=3,        # default is 6\n",
    "        #    ratio=5 )\n",
    "\n",
    "        # Add correlation coefficient as text\n",
    "        #g.fig.suptitle(f'Correlation (r) = {corr:.2f}' + f'\\n(n = {len(m)})', fontsize=10)\n",
    "        #g.set_axis_labels(s1,s2)\n",
    "        #g.ax_joint.set_ylim(0, 100) \n",
    "        #g.ax_joint.set_xlim(0, 100) \n",
    "        #g.fig.subplots_adjust(top=0.9)  # make room for suptitle\n",
    "        #comp = s1 + \"_vs_\" + s2\n",
    "        #g.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Correlation_plot_\" + comp + \".pseU.png\", transparent=True, dpi = 300)\n",
    "        #g.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Correlation_plot_\" + comp + \".pseU.pdf\", transparent=True, dpi = 300)\n",
    "        #plt.close()\n",
    "    else:\n",
    "        corr = float('nan')\n",
    "\n",
    "    result_rows.append({\n",
    "        'Sample1': s1,\n",
    "        'Sample2': s2,\n",
    "        'n_pos_cov>=10.Sample1': len(pos_cov1),\n",
    "        'n_pos_cov>=10.Sample2': len(pos_cov2),\n",
    "        'n_pseU_ratio>=10.Sample1': len(pos_mod1),\n",
    "        'n_pseU_ratio>=10.Sample2': len(pos_mod2),\n",
    "        'n_pos_cov>=10.Sample1_Sample2': len(joined_cov),\n",
    "        'n_pseU_ratio>=10.Sample1_joined_cov': len(df1_mod_joined),\n",
    "        'n_pseU_ratio>=10.Sample2_joined_cov': len(df2_mod_joined),\n",
    "        'n_pseU_ratio>=10.Sample1_Sample2': len(joined_cov_mod),\n",
    "        'mean_pseU_ratio.Sample1_joined_mod' : np.mean(mod1),\n",
    "        'median_pseU_ratio.Sample1_joined_mod' : np.median(mod1),\n",
    "        'mean_pseU_ratio.Sample2_joined_mod' : np.mean(mod2) ,\n",
    "        'median_pseU_ratio.Sample2_joined_mod' : np.median(mod2) ,\n",
    "        'Correlation_coeff_joined_cov_and_mod': corr\n",
    "    })\n",
    "    \n",
    "# Final result DataFrame\n",
    "comparison_df = pd.DataFrame(result_rows)\n",
    "comparison_df.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_TABLE_r2_values_all_pairwise_comps_pseU.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# CREATE COUNTS PLOTS AS BAR PLOT "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pseu_all_samples.loc[pseu_all_samples[\"coverage\"] >= 10, \"LABEL\"] = \"cov >= 10 reads\"\n",
    "pseu_all_samples.loc[(pseu_all_samples[\"coverage\"] >= 10) & (pseu_all_samples[\"modRatio\"] >= 10), \"LABEL\"] = \"pseU >= 10%\"\n",
    "pseu_all_samples_cov10 = pseu_all_samples[pseu_all_samples[\"coverage\"] >= 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "total = pseu_all_samples_cov10.groupby('sample').size().reset_index(name=\"total_count\")\n",
    "covered = pseu_all_samples_cov10[pseu_all_samples_cov10.LABEL == 'cov >= 10 reads'].groupby(\"sample\").size().reset_index(name=\"covered_count\")\n",
    "pseu = pseu_all_samples_cov10[pseu_all_samples_cov10.LABEL == 'pseU >= 10%'].groupby(\"sample\").size().reset_index(name=\"pseu_count\")\n",
    "# Merge all counts\n",
    "df = total.merge(covered, on='sample', how='left').merge(pseu, on='sample', how='left')\n",
    "df.fillna(0, inplace=True)  # fill missing values with 0\n",
    "# Calculate percentages\n",
    "df['pseu_pct'] = df['pseu_count'] / df['covered_count'] * 100\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors_input = {\n",
    "  \"RNA004_HEK293T_A\" : \"#EE6C4D\",\n",
    "  \"RNA004_HEK293T_B\" : \"#E69581\",\n",
    "  \"RNA004_blood_1\" :\"#0B3954\",\n",
    "  \"RNA004_blood_IVT_1\" :\"#8b5cf6\",\n",
    "  \"RNA004_UHRR_1\" : \"#098C9A\",\n",
    "  \"RNA004_UHRR_2\" : \"#26AFBE\",\n",
    "  \"RNA004_UHRR_3\" : \"#93C6CC\", \n",
    "  \"RNA004_blood_2\" : \"#177BB5\",\n",
    "  \"RNA004_blood_3\" : \"#38A6E5\",\n",
    "  \"RNA004_blood_IVT_3\" : \"#BA9EFA\",\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TOTAL COUNTS WITH COVERAGE >= 10 AND MODRATIO >= 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5, 3))\n",
    "sns.barplot(data = df, hue = \"sample\", x=\"sample\", edgecolor =\"black\", y=\"pseu_count\",saturation=1, palette = colors_input)\n",
    "ax.spines['left'].set_color('black')   # Left axis line\n",
    "ax.spines['bottom'].set_color('black') \n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation = 45, ha = \"right\", rotation_mode = \"anchor\")\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "for p in ax.patches:\n",
    "    height = p.get_height()\n",
    "    if not np.isnan(height):  # avoid labeling bars with NaN height\n",
    "        ax.text(\n",
    "            x=p.get_x() + p.get_width() / 2,\n",
    "            y=height + 1,  # small offset above bar\n",
    "            s=f\"  {int(height)}\",  # or f\"{height:.1f}\" for float\n",
    "            ha=\"center\",\n",
    "            va=\"bottom\",\n",
    "            rotation=90,\n",
    "            fontsize=9,\n",
    "            color = \"gray\"\n",
    "        )\n",
    "ax.set_ylabel(\"Total pseU sites (frequency >= 10)\")\n",
    "ax.set_xlabel(\"\")\n",
    "i = \"all_samples\"\n",
    "fig.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE4/PAPER_pseU_count_\" + i + \"_RNA004_tissue_bar.png\", bbox_inches='tight', dpi=300)\n",
    "fig.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE4/PAPER_pseU_count_\" + i + \"_RNA004_tissue_bar.pdf\", bbox_inches='tight', dpi=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## RMBASE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df = pd.DataFrame()\n",
    "\n",
    "for s in pseu_all_samples[\"sample\"].drop_duplicates():\n",
    "    print(s)\n",
    "    df_sub = pseu_all_samples[pseu_all_samples[\"sample\"] == s] \n",
    "    l_mod_sub = rmbase_sites[rmbase_sites[\"pos\"].isin(df_sub.pos)]\n",
    "    df_sig = df_sub[df_sub[\"modRatio\"] >= 10]\n",
    "    \n",
    "    num_l_mod = df_sig[df_sig[\"pos\"].isin(l_mod_sub.pos)]\n",
    "    print(len(num_l_mod)/len(l_mod_sub))\n",
    "    f = pd.DataFrame({\"sample\" : [s],\n",
    "                      \"num_cov_sites\": [len(df_sub)],\n",
    "                      \"num_sig_sites\": [len(df_sig)], \n",
    "                      \"num_RMBase_cov\" : [len(l_mod_sub)],\n",
    "                      \"perc_RMBase_mod\" : [len(num_l_mod)/len(l_mod_sub)*100]})\n",
    "    new_df = pd.concat([new_df, f], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "plt.figure(figsize=(5, 4))\n",
    "\n",
    "\n",
    "ax = sns.barplot(x='sample', y='perc_RMBase_mod', data=new_df, hue = \"sample\", palette=colors_input, label='pseU sites', edgecolor = \"black\", order = sorted(df[\"sample\"].drop_duplicates()))\n",
    "\n",
    "bottom_bar = mpatches.Patch(color='#0b3954', label='pseU sites')\n",
    "\n",
    "# Set labels\n",
    "plt.ylabel('Percentage of RMBase pseU sites [%]')\n",
    "plt.xlabel(\"\")\n",
    "df[\"sample_flat\"] = df[\"sample\"].str.replace(\"RNA004_\", \"\")\n",
    "\n",
    "for p in ax.patches:\n",
    "    height = p.get_height()\n",
    "    if not pd.isna(height):  # skip missing values\n",
    "        ax.text(\n",
    "            p.get_x() + 0.3 + p.get_width() / 2,\n",
    "            height + 0.5,  # slightly above the bar\n",
    "            f'{height:.1f}%',\n",
    "            ha='center',\n",
    "            va='bottom',rotation = 45,\n",
    "            fontsize=8\n",
    "        )\n",
    "\n",
    "plt.xticks(ticks = sorted(df[\"sample\"].drop_duplicates()), labels = sorted(df[\"sample_flat\"].drop_duplicates()),rotation = 45, ha='right')\n",
    "plt.tight_layout()\n",
    "\n",
    "# show the graph\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/SUPP_RMBase_pseu_amount_from_cov_sites.pdf\", transparent = True, dpi = 300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TAVAKOLI SITES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_df_tava = pd.DataFrame()\n",
    "\n",
    "for s in pseu_all_samples[\"sample\"].drop_duplicates():\n",
    "    print(s)\n",
    "    df_sub = pseu_all_samples[pseu_all_samples[\"sample\"] == s] \n",
    "    l_mod_sub2 = tava_sites[tava_sites[\"pos\"].isin(df_sub.pos)]\n",
    "    df_sig = df_sub[df_sub[\"modRatio\"] >= 10]\n",
    "    \n",
    "    num_l_mod = df_sig[df_sig[\"pos\"].isin(l_mod_sub2.pos)]\n",
    "    print(len(num_l_mod)/len(l_mod_sub2))\n",
    "    f = pd.DataFrame({\"sample\" : [s],\n",
    "                      \"num_cov_sites\": [len(df_sub)],\n",
    "                      \"num_sig_sites\": [len(df_sig)], \n",
    "                      \"num_RMBase_cov\" : [len(l_mod_sub2)],\n",
    "                      \"perc_RMBase_mod\" : [len(num_l_mod)/len(l_mod_sub2)*100]})\n",
    "    new_df_tava = pd.concat([new_df_tava, f], axis = 0)\n",
    "plt.figure(figsize=(5, 4))\n",
    "\n",
    "\n",
    "ax = sns.barplot(x='sample', y='perc_RMBase_mod', data=new_df_tava, saturation = 1, alpha = 1, hue = \"sample\", palette=colors_input, label='pseU sites', edgecolor = \"black\", order = sorted(df[\"sample\"].drop_duplicates()))\n",
    "\n",
    "bottom_bar = mpatches.Patch(color='#0b3954', label='pseU sites')\n",
    "\n",
    "# Set labels\n",
    "plt.ylabel('Percentage of Tavakoli pseU sites [%]')\n",
    "plt.xlabel(\"\")\n",
    "df[\"sample_flat\"] = df[\"sample\"].str.replace(\"RNA004_\", \"\")\n",
    "\n",
    "for p in ax.patches:\n",
    "    height = p.get_height()\n",
    "    if not pd.isna(height):  # skip missing values\n",
    "        ax.text(\n",
    "            p.get_x() + 0.3 + p.get_width() / 2,\n",
    "            height + 0.5,  # slightly above the bar\n",
    "            f'{height:.1f}%',\n",
    "            ha='center',\n",
    "            va='bottom',rotation = 45,\n",
    "            fontsize=8\n",
    "        )\n",
    "\n",
    "plt.xticks(ticks = sorted(df[\"sample\"].drop_duplicates()), labels = sorted(df[\"sample_flat\"].drop_duplicates()),rotation = 45, ha='right')\n",
    "plt.tight_layout()\n",
    "\n",
    "# show the graph\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/SUPP_Tavakoli_pseu_amount_from_cov_sites.pdf\", transparent = True, dpi = 300)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# PSEU OVERLAPS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def generate_intersections_table(input_dict):\n",
    "    # Prepare results list\n",
    "    results = []\n",
    "    data = input_dict\n",
    "    # Convert keys to a list to allow combinations\n",
    "    keys = list(data.keys())\n",
    "\n",
    "    # Loop through all non-empty combinations of the keys\n",
    "    for r in range(1, len(keys) + 1):\n",
    "        for combo in combinations(keys, r):\n",
    "            # Start with the first set in the combination\n",
    "            intersection = data[combo[0]]\n",
    "            for key in combo[1:]:\n",
    "                intersection = intersection & data[key]\n",
    "            \n",
    "            # Subtract elements shared with any *other* sets not in the combo\n",
    "            other_keys = set(keys) - set(combo)\n",
    "            for key in other_keys:\n",
    "                intersection = intersection - data[key]\n",
    "            \n",
    "            if len(keys) == 3:\n",
    "                results.append({\n",
    "                    \"Intersection\": \" ∩ \".join(combo),\n",
    "                    \"Count\": len(intersection),\n",
    "                    \"n_\" + keys[0] : len(input_dict[keys[0]]),\n",
    "                    \"n_\" + keys[1] : len(input_dict[keys[1]]),\n",
    "                    \"n_\" + keys[2] : len(input_dict[keys[2]]),\n",
    "            }) \n",
    "            else:\n",
    "                results.append({\n",
    "                    \"Intersection\": \" ∩ \".join(combo),\n",
    "                    \"Count\": len(intersection),\n",
    "                    \"n_\" + keys[0] : len(input_dict[keys[0]]),\n",
    "                    \"n_\" + keys[1] : len(input_dict[keys[1]])\n",
    "            })   \n",
    "\n",
    "    # Create DataFrame\n",
    "    df = pd.DataFrame(results)\n",
    "    return(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_pseu_cov10_freq10 = p_pseu_cov10[p_pseu_cov10[\"modRatio\"] >= 10]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "single_dorado1 = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_blood\")]\n",
    "single_dorado2 = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_S5_DRS\")]\n",
    "single_dorado3 = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_S6_DRS\")]\n",
    "single_dorado_IVT = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_S6_IVT_merged\")]\n",
    "\n",
    "# EXTRACT SITES COVERED BY IVT\n",
    "single_dorado1_subIVT1 = single_dorado1[single_dorado1[\"pos\"].isin(single_dorado_IVT.pos)]\n",
    "single_dorado2_subIVT1 = single_dorado2[single_dorado2[\"pos\"].isin(single_dorado_IVT.pos)]\n",
    "single_dorado3_subIVT1 = single_dorado3[single_dorado3[\"pos\"].isin(single_dorado_IVT.pos)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SIGNIFICANT SITES PER SAMPLE (ALL SITES)\n",
    "single_dorado1_freq = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_blood\") & (p_pseu_cov10[\"modRatio\"] >= 10)]\n",
    "single_dorado2_freq = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_S5_DRS\") & (p_pseu_cov10[\"modRatio\"] >= 10)]\n",
    "single_dorado3_freq = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_S6_DRS\") & (p_pseu_cov10[\"modRatio\"] >= 10)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "single_dorado_freq_IVT = p_pseu_cov10[(p_pseu_cov10[\"sample\"] == \"RNA004_S6_IVT_merged\") & (p_pseu_cov10[\"modRatio\"] >= 10)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SITES COVERED BY IVT AND BLOOD SAMPLE AND SIGNIFICANT IN BLOOD SAMPLE\n",
    "single_dorado1_subIVT1_freq = single_dorado1_subIVT1[single_dorado1_subIVT1[\"modRatio\"] >= 10]\n",
    "single_dorado2_subIVT1_freq = single_dorado2_subIVT1[single_dorado2_subIVT1[\"modRatio\"] >= 10]\n",
    "single_dorado3_subIVT1_freq = single_dorado3_subIVT1[single_dorado3_subIVT1[\"modRatio\"] >= 10]\n",
    "single_dorado_IVT_freq = single_dorado_IVT[single_dorado_IVT[\"modRatio\"] >= 10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overlap between replicates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SITES COVERED WITH AT LEAST 10 READS\n",
    "DORADO_RNA004_blood1 = single_dorado1\n",
    "DORADO_RNA004_blood2 = single_dorado2\n",
    "DORADO_RNA004_blood3 = single_dorado3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SIGNIFICANT SITES PER SAMPLE COVERED BY ALL BLOOD SAMPLES\n",
    "DORADO_RNA004_blood1_freq = DORADO_RNA004_blood1.loc[(DORADO_RNA004_blood1[\"modRatio\"] >= 10) & (DORADO_RNA004_blood1.pos.isin(DORADO_RNA004_blood2.pos) & DORADO_RNA004_blood1.pos.isin(DORADO_RNA004_blood3.pos)), :]\n",
    "DORADO_RNA004_blood2_freq = DORADO_RNA004_blood2.loc[(DORADO_RNA004_blood2[\"modRatio\"] >= 10) & (DORADO_RNA004_blood2.pos.isin(DORADO_RNA004_blood1.pos) & DORADO_RNA004_blood2.pos.isin(DORADO_RNA004_blood3.pos)), :]\n",
    "DORADO_RNA004_blood3_freq = DORADO_RNA004_blood3.loc[(DORADO_RNA004_blood3[\"modRatio\"] >= 10) & (DORADO_RNA004_blood3.pos.isin(DORADO_RNA004_blood1.pos) & DORADO_RNA004_blood3.pos.isin(DORADO_RNA004_blood2.pos)), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO_RNA004_blood1_cov = DORADO_RNA004_blood1.loc[(DORADO_RNA004_blood1.pos.isin(DORADO_RNA004_blood2.pos) & DORADO_RNA004_blood1.pos.isin(DORADO_RNA004_blood3.pos)), :]\n",
    "DORADO_RNA004_blood2_cov = DORADO_RNA004_blood2.loc[(DORADO_RNA004_blood2.pos.isin(DORADO_RNA004_blood1.pos) & DORADO_RNA004_blood2.pos.isin(DORADO_RNA004_blood3.pos)), :]\n",
    "DORADO_RNA004_blood3_cov = DORADO_RNA004_blood3.loc[(DORADO_RNA004_blood3.pos.isin(DORADO_RNA004_blood1.pos) & DORADO_RNA004_blood3.pos.isin(DORADO_RNA004_blood2.pos)), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_dict2 = {\"blood_1\" : set(DORADO_RNA004_blood1_freq.pos), \"blood_2\" : set(DORADO_RNA004_blood2_freq.pos), \"blood_3\" : set(DORADO_RNA004_blood3_freq.pos)}\n",
    "df_int = generate_intersections_table(my_dict2)\n",
    "df_int.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_Intersections_blood_RNA004_pseU_joined_cov_sites.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "\n",
    "sets = list(my_dict2.values())\n",
    "\n",
    "labels = list(my_dict2.keys())\n",
    "labels[0] = labels[0] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood1_freq.pos)))  + \")\"\n",
    "labels[1] = labels[1] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood2_freq.pos))) + \")\"\n",
    "labels[2] = labels[2] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood3_freq.pos)))  + \")\"\n",
    "\n",
    "# Create Venn diagram\n",
    "venn = venn3(subsets=sets, set_labels=labels)\n",
    "\n",
    "\n",
    "venn.get_patch_by_id('100').set_color(\"#1F97DD\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('010').set_color(\"#53ABDF\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('001').set_color(\"#84BADA\")   # mAFiA\n",
    "venn.get_patch_by_id('110').set_color(\"#176796\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('101').set_color(\"#4E7F9B\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('111').set_color(\"#0B3954\")   # mAFiA\n",
    "venn.get_patch_by_id('011').set_color(\"#5B839B\")   # mAFiA\n",
    "\n",
    "venn.get_label_by_id('100').set_color(\"white\")   # Dorado only\n",
    "venn.get_label_by_id('010').set_color(\"white\")   # m6ABasecaller only\n",
    "venn.get_label_by_id('001').set_color(\"white\")   # mAFiA only\n",
    "venn.get_label_by_id('110').set_color(\"white\")   # Dorado + m6ABasecaller\n",
    "venn.get_label_by_id('101').set_color(\"white\")   # Dorado + mAFiA\n",
    "venn.get_label_by_id('111').set_color(\"white\")   # All three\n",
    "venn.get_label_by_id('011').set_color(\"white\")   # All three\n",
    "\n",
    "venn.get_label_by_id('100').set_fontsize(7)\n",
    "venn.get_label_by_id('010').set_fontsize(7)\n",
    "venn.get_label_by_id('001').set_fontsize(7)\n",
    "venn.get_label_by_id('110').set_fontsize(7)\n",
    "venn.get_label_by_id('101').set_fontsize(7)\n",
    "venn.get_label_by_id('111').set_fontsize(17)\n",
    "venn.get_label_by_id('011').set_fontsize(7)\n",
    "\n",
    "\n",
    "for patch in venn.patches:\n",
    "    if patch:  # some regions might be empty (None)\n",
    "        patch.set_edgecolor(\"black\")\n",
    "        patch.set_linewidth(2)  # adjust thickness as needed\n",
    "        patch.set_alpha(1)\n",
    "        print(patch.get_label())\n",
    "\n",
    "\n",
    "\n",
    "# Change set label font size (those around the circles)\n",
    "for text in venn.set_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(20)  # Set your desired font size here\n",
    "\n",
    "\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_pseU.png\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_pseU.pdf\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overlap without IVT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO_RNA004_blood1_freq_IVT = DORADO_RNA004_blood1_freq[~DORADO_RNA004_blood1_freq.pos.isin(single_dorado_freq_IVT.pos)]\n",
    "DORADO_RNA004_blood2_freq_IVT = DORADO_RNA004_blood2_freq[~DORADO_RNA004_blood2_freq.pos.isin(single_dorado_freq_IVT.pos)]\n",
    "DORADO_RNA004_blood3_freq_IVT = DORADO_RNA004_blood3_freq[~DORADO_RNA004_blood3_freq.pos.isin(single_dorado_freq_IVT.pos)]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_dict2 = {\"blood_1\" : set(DORADO_RNA004_blood1_freq_IVT.pos), \"blood_2\" : set(DORADO_RNA004_blood2_freq_IVT.pos), \"blood_3\" : set(DORADO_RNA004_blood3_freq_IVT.pos)}\n",
    "df_int = generate_intersections_table(my_dict2)\n",
    "df_int.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_Intersections_blood_RNA004_pseU_joined_cov_sites_minusIVT.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "sets = list(my_dict2.values())\n",
    "\n",
    "labels = list(my_dict2.keys())\n",
    "labels[0] = labels[0] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood1_freq_IVT.pos)))  + \")\"\n",
    "labels[1] = labels[1] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood2_freq_IVT.pos))) + \")\"\n",
    "labels[2] = labels[2] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood3_freq_IVT.pos)))  + \")\"\n",
    "\n",
    "# Create Venn diagram\n",
    "venn = venn3(subsets=sets, set_labels=labels)\n",
    "\n",
    "\n",
    "venn.get_patch_by_id('100').set_color(\"#1F97DD\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('010').set_color(\"#53ABDF\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('001').set_color(\"#84BADA\")   # mAFiA\n",
    "venn.get_patch_by_id('110').set_color(\"#176796\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('101').set_color(\"#4E7F9B\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('111').set_color(\"#0B3954\")   # mAFiA\n",
    "venn.get_patch_by_id('011').set_color(\"#5B839B\")   # mAFiA\n",
    "\n",
    "#A25B4F\n",
    "# Optional: Set transpare\n",
    "\n",
    "venn.get_label_by_id('100').set_color(\"white\")   # Dorado only\n",
    "venn.get_label_by_id('010').set_color(\"white\")   # m6ABasecaller only\n",
    "venn.get_label_by_id('001').set_color(\"white\")   # mAFiA only\n",
    "venn.get_label_by_id('110').set_color(\"white\")   # Dorado + m6ABasecaller\n",
    "venn.get_label_by_id('101').set_color(\"white\")   # Dorado + mAFiA\n",
    "venn.get_label_by_id('111').set_color(\"white\")   # All three\n",
    "venn.get_label_by_id('011').set_color(\"white\")   # All three\n",
    "\n",
    "venn.get_label_by_id('100').set_fontsize(7)\n",
    "venn.get_label_by_id('010').set_fontsize(7)\n",
    "venn.get_label_by_id('001').set_fontsize(7)\n",
    "venn.get_label_by_id('110').set_fontsize(7)\n",
    "venn.get_label_by_id('101').set_fontsize(7)\n",
    "venn.get_label_by_id('111').set_fontsize(10)\n",
    "venn.get_label_by_id('011').set_fontsize(7)\n",
    "\n",
    "\n",
    "for patch in venn.patches:\n",
    "    if patch:  # some regions might be empty (None)\n",
    "        patch.set_edgecolor(\"black\")\n",
    "        patch.set_linewidth(2)  # adjust thickness as needed\n",
    "        patch.set_alpha(1)\n",
    "        print(patch.get_label())\n",
    "\n",
    "\n",
    "\n",
    "# Change set label font size (those around the circles)\n",
    "for text in venn.set_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(20)  # Set your desired font size here\n",
    "\n",
    "\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_without_IVT_pseU.png\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_without_IVT_pseU.pdf\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## OVERLAP OF ALL SITES "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO_RNA004_blood1_freq_all = DORADO_RNA004_blood1.loc[(DORADO_RNA004_blood1[\"modRatio\"] >= 10), :]\n",
    "DORADO_RNA004_blood2_freq_all = DORADO_RNA004_blood2.loc[(DORADO_RNA004_blood2[\"modRatio\"] >= 10), :]\n",
    "DORADO_RNA004_blood3_freq_all = DORADO_RNA004_blood3.loc[(DORADO_RNA004_blood3[\"modRatio\"] >= 10), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "my_dict2 = {\"blood_1\" : set(DORADO_RNA004_blood1_freq_all.pos), \"blood_2\" : set(DORADO_RNA004_blood2_freq_all.pos), \"blood_3\" : set(DORADO_RNA004_blood3_freq_all.pos)}\n",
    "\n",
    "df_int = generate_intersections_table(my_dict2)\n",
    "df_int.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_Intersections_blood_RNA004_pseU_all_sites.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "sets = list(my_dict2.values())\n",
    "\n",
    "labels = list(my_dict2.keys())\n",
    "labels[0] = labels[0] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood1_freq_all.pos)))  + \")\"\n",
    "labels[1] = labels[1] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood2_freq_all.pos))) + \")\"\n",
    "labels[2] = labels[2] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood3_freq_all.pos)))  + \")\"\n",
    "\n",
    "# Create Venn diagram\n",
    "venn = venn3(subsets=sets, set_labels=labels)\n",
    "\n",
    "\n",
    "venn.get_patch_by_id('100').set_color(\"#1F97DD\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('010').set_color(\"#53ABDF\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('001').set_color(\"#84BADA\")   # mAFiA\n",
    "venn.get_patch_by_id('110').set_color(\"#176796\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('101').set_color(\"#4E7F9B\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('111').set_color(\"#0B3954\")   # mAFiA\n",
    "venn.get_patch_by_id('011').set_color(\"#5B839B\")   # mAFiA\n",
    "\n",
    "#A25B4F\n",
    "# Optional: Set transpare\n",
    "\n",
    "venn.get_label_by_id('100').set_color(\"white\")   # Dorado only\n",
    "venn.get_label_by_id('010').set_color(\"white\")   # m6ABasecaller only\n",
    "venn.get_label_by_id('001').set_color(\"white\")   # mAFiA only\n",
    "venn.get_label_by_id('110').set_color(\"white\")   # Dorado + m6ABasecaller\n",
    "venn.get_label_by_id('101').set_color(\"white\")   # Dorado + mAFiA\n",
    "venn.get_label_by_id('111').set_color(\"white\")   # All three\n",
    "venn.get_label_by_id('011').set_color(\"white\")   # All three\n",
    "\n",
    "venn.get_label_by_id('100').set_fontsize(7)\n",
    "venn.get_label_by_id('010').set_fontsize(7)\n",
    "venn.get_label_by_id('001').set_fontsize(7)\n",
    "venn.get_label_by_id('110').set_fontsize(7)\n",
    "venn.get_label_by_id('101').set_fontsize(7)\n",
    "venn.get_label_by_id('111').set_fontsize(10)\n",
    "venn.get_label_by_id('011').set_fontsize(7)\n",
    "\n",
    "\n",
    "for patch in venn.patches:\n",
    "    if patch:  # some regions might be empty (None)\n",
    "        patch.set_edgecolor(\"black\")\n",
    "        patch.set_linewidth(2)  # adjust thickness as needed\n",
    "        patch.set_alpha(1)\n",
    "        print(patch.get_label())\n",
    "\n",
    "\n",
    "\n",
    "# Change set label font size (those around the circles)\n",
    "for text in venn.set_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(20)  # Set your desired font size here\n",
    "\n",
    "\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_all_sites_pseu.png\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_all_sites_pseu.pdf\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OVERLAP OF ALL SITES MINUS IVT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO_RNA004_blood1_freq_IVT_all = DORADO_RNA004_blood1_freq_all[~DORADO_RNA004_blood1_freq_all.pos.isin(single_dorado_freq_IVT.pos)]\n",
    "DORADO_RNA004_blood2_freq_IVT_all = DORADO_RNA004_blood2_freq_all[~DORADO_RNA004_blood2_freq_all.pos.isin(single_dorado_freq_IVT.pos)]\n",
    "DORADO_RNA004_blood3_freq_IVT_all = DORADO_RNA004_blood3_freq_all[~DORADO_RNA004_blood3_freq_all.pos.isin(single_dorado_freq_IVT.pos)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "my_dict2 = {\"blood_1\" : set(DORADO_RNA004_blood1_freq_IVT_all.pos), \"blood_2\" : set(DORADO_RNA004_blood2_freq_IVT_all.pos), \"blood_3\" : set(DORADO_RNA004_blood3_freq_IVT_all.pos)}\n",
    "\n",
    "df_int = generate_intersections_table(my_dict2)\n",
    "df_int.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_Intersections_blood_RNA004_pseU_all_sites_minusIVT.tsv\", sep = \"\\t\", index = False)\n",
    "\n",
    "sets = list(my_dict2.values())\n",
    "\n",
    "labels = list(my_dict2.keys())\n",
    "labels[0] = labels[0] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood1_freq_IVT_all.pos)))  + \")\"\n",
    "labels[1] = labels[1] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood2_freq_IVT_all.pos))) + \")\"\n",
    "labels[2] = labels[2] + \"\\n(n=\" + str(len(set(DORADO_RNA004_blood3_freq_IVT_all.pos)))  + \")\"\n",
    "\n",
    "# Create Venn diagram\n",
    "venn = venn3(subsets=sets, set_labels=labels)\n",
    "\n",
    "venn.get_patch_by_id('100').set_color(\"#1F97DD\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('010').set_color(\"#53ABDF\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('001').set_color(\"#84BADA\")   # mAFiA\n",
    "venn.get_patch_by_id('110').set_color(\"#176796\")   # Dorado@v5.0.0\n",
    "venn.get_patch_by_id('101').set_color(\"#4E7F9B\")   # m6ABasecaller\n",
    "venn.get_patch_by_id('111').set_color(\"#0B3954\")   # mAFiA\n",
    "venn.get_patch_by_id('011').set_color(\"#5B839B\")   # mAFiA\n",
    "\n",
    "venn.get_label_by_id('100').set_color(\"white\")   # Dorado only\n",
    "venn.get_label_by_id('010').set_color(\"white\")   # m6ABasecaller only\n",
    "venn.get_label_by_id('001').set_color(\"white\")   # mAFiA only\n",
    "venn.get_label_by_id('110').set_color(\"white\")   # Dorado + m6ABasecaller\n",
    "venn.get_label_by_id('101').set_color(\"white\")   # Dorado + mAFiA\n",
    "venn.get_label_by_id('111').set_color(\"white\")   # All three\n",
    "venn.get_label_by_id('011').set_color(\"white\")   # All three\n",
    "\n",
    "venn.get_label_by_id('100').set_fontsize(7)\n",
    "venn.get_label_by_id('010').set_fontsize(7)\n",
    "venn.get_label_by_id('001').set_fontsize(7)\n",
    "venn.get_label_by_id('110').set_fontsize(7)\n",
    "venn.get_label_by_id('101').set_fontsize(7)\n",
    "venn.get_label_by_id('111').set_fontsize(10)\n",
    "venn.get_label_by_id('011').set_fontsize(7)\n",
    "\n",
    "\n",
    "for patch in venn.patches:\n",
    "    if patch:  # some regions might be empty (None)\n",
    "        patch.set_edgecolor(\"black\")\n",
    "        patch.set_linewidth(2)  # adjust thickness as needed\n",
    "        patch.set_alpha(1)\n",
    "        print(patch.get_label())\n",
    "\n",
    "\n",
    "\n",
    "# Change set label font size (those around the circles)\n",
    "for text in venn.set_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(20)  # Set your desired font size here\n",
    "\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_all_sites_minusIVT.png\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_blood_RNA004_blood_cov10_freq10_replicates_all_sites_minusIVT.pdf\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# OVERLAP OF CELL-LINES"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PSEU UHRR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO_RNA004_UHRR1 = p_cell_pseu_cov10[p_cell_pseu_cov10[\"sample\"] == \"RNA004_UHRR_1\"]\n",
    "DORADO_RNA004_UHRR2 = p_cell_pseu_cov10[p_cell_pseu_cov10[\"sample\"] == \"RNA004_UHRR_2\"]\n",
    "DORADO_RNA004_UHRR3 = p_cell_pseu_cov10[p_cell_pseu_cov10[\"sample\"] == \"RNA004_UHRR_3\"]\n",
    "DORADO_RNA004_UHRR1_freq = DORADO_RNA004_UHRR1.loc[(DORADO_RNA004_UHRR1[\"modRatio\"] >= 10) & (DORADO_RNA004_UHRR1.pos.isin(DORADO_RNA004_UHRR2.pos) & DORADO_RNA004_UHRR1.pos.isin(DORADO_RNA004_UHRR3.pos)), :]\n",
    "DORADO_RNA004_UHRR2_freq = DORADO_RNA004_UHRR2.loc[(DORADO_RNA004_UHRR2[\"modRatio\"] >= 10) & (DORADO_RNA004_UHRR2.pos.isin(DORADO_RNA004_UHRR1.pos) & DORADO_RNA004_UHRR2.pos.isin(DORADO_RNA004_UHRR3.pos)), :]\n",
    "DORADO_RNA004_UHRR3_freq = DORADO_RNA004_UHRR3.loc[(DORADO_RNA004_UHRR3[\"modRatio\"] >= 10) & (DORADO_RNA004_UHRR3.pos.isin(DORADO_RNA004_UHRR1.pos) & DORADO_RNA004_UHRR3.pos.isin(DORADO_RNA004_UHRR2.pos)), :]\n",
    "overlapping_cov_sites_cell = DORADO_RNA004_UHRR2[(DORADO_RNA004_UHRR2.pos.isin(DORADO_RNA004_UHRR3.pos)) & (DORADO_RNA004_UHRR2.pos.isin(DORADO_RNA004_UHRR1.pos))]\n",
    "len(overlapping_cov_sites_cell)\n",
    "\n",
    "my_dict2 = {\"UHRR_1\" : set(DORADO_RNA004_UHRR1_freq.pos), \"UHRR_2\" : set(DORADO_RNA004_UHRR2_freq.pos), \"UHRR_3\" : set(DORADO_RNA004_UHRR3_freq.pos)}\n",
    "sets = list(my_dict2.values())\n",
    "\n",
    "\n",
    "df_int = generate_intersections_table(my_dict2)\n",
    "df_int[\"joined sites (cov >= 10)\"] = len(overlapping_cov_sites_cell)\n",
    "df_int.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_RNA004_INTERSECTION_UHRR_joined_cov_sites_pseU.tsv\", sep = \"\\t\", index = False)\n",
    "labels = list(my_dict2.keys())\n",
    "labels[0] = labels[0] + \"\\n(n=\" + str(len(set(DORADO_RNA004_UHRR1_freq.pos)))  + \")\"\n",
    "labels[1] = labels[1] + \"\\n(n=\" + str(len(set(DORADO_RNA004_UHRR2_freq.pos))) + \")\"\n",
    "labels[2] = labels[2] + \"\\n(n=\" + str(len(set(DORADO_RNA004_UHRR3_freq.pos)))  + \")\"\n",
    "\n",
    "# Create Venn diagram\n",
    "venn = venn3(subsets=sets, set_labels=labels)\n",
    "\n",
    "venn.get_patch_by_id('100').set_color(\"#4DAFB9\")   \n",
    "venn.get_patch_by_id('010').set_color(\"#5DBBC5\")   \n",
    "venn.get_patch_by_id('001').set_color(\"#638D92\")   \n",
    "venn.get_patch_by_id('110').set_color(\"#10C1D4\")   \n",
    "venn.get_patch_by_id('101').set_color(\"#5ADAE9\")   \n",
    "venn.get_patch_by_id('111').set_color(\"#098C9A\")  \n",
    "venn.get_patch_by_id('011').set_color(\"#8AD8E0\")   \n",
    "\n",
    "\n",
    "venn.get_label_by_id('100').set_color(\"white\")   \n",
    "venn.get_label_by_id('010').set_color(\"white\")   \n",
    "venn.get_label_by_id('001').set_color(\"white\")  \n",
    "venn.get_label_by_id('110').set_color(\"white\")   \n",
    "venn.get_label_by_id('101').set_color(\"white\")   \n",
    "venn.get_label_by_id('111').set_color(\"white\")   \n",
    "venn.get_label_by_id('011').set_color(\"white\")  \n",
    "\n",
    "venn.get_label_by_id('100').set_fontsize(7)\n",
    "venn.get_label_by_id('010').set_fontsize(7)\n",
    "venn.get_label_by_id('001').set_fontsize(7)\n",
    "venn.get_label_by_id('110').set_fontsize(7)\n",
    "venn.get_label_by_id('101').set_fontsize(7)\n",
    "venn.get_label_by_id('111').set_fontsize(17)\n",
    "venn.get_label_by_id('011').set_fontsize(7)\n",
    "\n",
    "for patch in venn.patches:\n",
    "    if patch:  \n",
    "        patch.set_edgecolor(\"black\")\n",
    "        patch.set_linewidth(2) \n",
    "        patch.set_alpha(1)\n",
    "        print(patch.get_label())\n",
    "\n",
    "for text in venn.set_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(20)  \n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_UHRR_RNA004_cov10_freq10_replicates_pseU.png\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_UHRR_RNA004_cov10_freq10_replicates_pseU.pdf\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## PSEU HEK293T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "DORADO_RNA004_HEK_A = p_cell_pseu_cov10[p_cell_pseu_cov10[\"sample\"] == \"RNA004_A\"]\n",
    "DORADO_RNA004_HEK_B = p_cell_pseu_cov10[p_cell_pseu_cov10[\"sample\"] == \"RNA004_B\"]\n",
    "DORADO_RNA004_HEK_A_freq = DORADO_RNA004_HEK_A.loc[(DORADO_RNA004_HEK_A[\"modRatio\"] >= 10) & (DORADO_RNA004_HEK_A.pos.isin(DORADO_RNA004_HEK_B.pos)), :]\n",
    "DORADO_RNA004_HEK_B_freq = DORADO_RNA004_HEK_B.loc[(DORADO_RNA004_HEK_B[\"modRatio\"] >= 10) & (DORADO_RNA004_HEK_B.pos.isin(DORADO_RNA004_HEK_A.pos)), :]\n",
    "overlapping_cov_sites_cell = DORADO_RNA004_HEK_A[(DORADO_RNA004_HEK_A.pos.isin(DORADO_RNA004_HEK_B.pos))]\n",
    "len(overlapping_cov_sites_cell)\n",
    "\n",
    "my_dict2 = {\"HEK293T_A\" : set(DORADO_RNA004_HEK_A_freq.pos), \"HEK293T_B\" : set(DORADO_RNA004_HEK_B_freq.pos)}\n",
    "sets = list(my_dict2.values())\n",
    "df_int = generate_intersections_table(my_dict2)\n",
    "df_int[\"joined sites (cov >= 10)\"] = len(overlapping_cov_sites_cell)\n",
    "df_int.to_csv(\"/home/awiercze/NEW_RNA004_PLOTS/TABLES/PAPER_RNA004_INTERSECTION_HEK293T_joined_cov_sites_pseU.tsv\", sep = \"\\t\", index = False)\n",
    "labels = list(my_dict2.keys())\n",
    "labels[0] = labels[0] + \"\\n(n=\" + str(len(set(DORADO_RNA004_HEK_A_freq.pos)))  + \")\"\n",
    "labels[1] = labels[1] + \"\\n(n=\" + str(len(set(DORADO_RNA004_HEK_B_freq.pos))) + \")\"\n",
    "\n",
    "\n",
    "# Create Venn diagram\n",
    "venn = venn2(subsets=sets, set_labels=labels)\n",
    "\n",
    "venn.get_patch_by_id('10').set_color(\"#F5BAAC\")   \n",
    "venn.get_patch_by_id('01').set_color(\"#EE6C4D\")   \n",
    "venn.get_patch_by_id('11').set_color(\"#EB9782\")   \n",
    "\n",
    "\n",
    "venn.get_label_by_id('10').set_color(\"white\")   \n",
    "venn.get_label_by_id('01').set_color(\"white\")   \n",
    "venn.get_label_by_id('11').set_color(\"white\")  \n",
    "\n",
    "venn.get_label_by_id('10').set_fontsize(7)\n",
    "venn.get_label_by_id('11').set_fontsize(17)\n",
    "venn.get_label_by_id('01').set_fontsize(7)\n",
    "\n",
    "for patch in venn.patches:\n",
    "    if patch:  \n",
    "        patch.set_edgecolor(\"black\")\n",
    "        patch.set_linewidth(2) \n",
    "        patch.set_alpha(1)\n",
    "        print(patch.get_label())\n",
    "\n",
    "\n",
    "\n",
    "for text in venn.set_labels:\n",
    "    if text:\n",
    "        text.set_fontsize(20)  \n",
    "\n",
    "\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_HEK293T_RNA004_cov10_freq10_replicates_pseU.png\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/Overlapping_covered_HEK293T_RNA004_cov10_freq10_replicates_pseU.pdf\", dpi=300, bbox_inches='tight', transparent=True)\n",
    "\n",
    "plt.show()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
