{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0f98e475",
   "metadata": {},
   "source": [
    "# PANEL 6\n",
    "\n",
    "**Input:** \n",
    "1. Basecalled and aligned bam files including pseU calls of HEK293T and oligo samples\n",
    "2. Basecalled and aligned bam files including m6A calls for 18S rRNA and blood samples (patient, healthy controls and IVT) \n",
    "\n",
    "**Output:** \n",
    "1. FIGURES 6E and 6G\n",
    "2. SUPPLEMENTARY: Modification probability plots for pseU calling of vectors, oligos and known gene site\n",
    "2. FIGURES 7F and 7G"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "174e2401",
   "metadata": {},
   "source": [
    "## Load packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8556f95",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pysam\n",
    "import seaborn as sns\n",
    "import matplotlib.pylab as plt\n",
    "import pyranges as pr"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fbeb074",
   "metadata": {},
   "source": [
    "## 1. Extract U>C mismatch percentages and modification calls from dorado for regions of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "89c90e54",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Extract U>C mismatch percentages\n",
    "def extract_modifcation_infos(i, out_df_mod, mod_pos, threshold, key = ('T', 0, 17802)):\n",
    "    # Check if modification present\n",
    "    if key not in i.modified_bases.keys():\n",
    "        return(out_df_mod)\n",
    "    # Extract modification information\n",
    "    mod_obj = i.modified_bases[key]\n",
    "    mod_info = [m for m in mod_obj if m[0] == mod_pos[0][0]]\n",
    "    # If no modification detected\n",
    "    if len(mod_info) == 0:\n",
    "        mod_info = [(mod_pos[0][0],0)]\n",
    "        # Set modification probability to 0\n",
    "        modprob = 0\n",
    "        #print(mod_pos)\n",
    "        # Check if mismatch was found (represented by reference base in small letters)\n",
    "        if mod_pos[0][2] in [\"t\", \"a\", \"g\", \"c\"]:\n",
    "            # Extract mismatch base\n",
    "            called_base = i.query_sequence[mod_pos[0][0]]\n",
    "            # If mismatch base == C, count for mismtach_C label\n",
    "            if called_base == \"C\":\n",
    "                out_df_mod[\"mismatch_C\"] += 1\n",
    "            else:\n",
    "                # If mismatch base == A or U or G, count for mismtach_AUG label\n",
    "                out_df_mod[\"mismatch_AUG\"] += 1\n",
    "        # If mod pos is empty -> deletion\n",
    "        elif mod_pos[0][0] == None:\n",
    "            out_df_mod[\"deletion\"] += 1\n",
    "        else:\n",
    "            out_df_mod[\"canonical\"] += 1\n",
    "    # If modification present\n",
    "    else:\n",
    "        # Reformate probability\n",
    "        modprob = (mod_info[0][1]+1)/256\n",
    "        # Check if modification passes threshold (p = 0.98)\n",
    "        if modprob < threshold:\n",
    "            out_df_mod[\"fail\"] += 1\n",
    "        else:\n",
    "            out_df_mod[\"modified\"] += 1\n",
    "    return(out_df_mod)\n",
    "\n",
    "# FUnction to extract the modification probabilites from DORADO\n",
    "def get_mod_probs(i, mod_pos, key = ('T', 0, 17802)):\n",
    "    modprob = None\n",
    "    if key in i.modified_bases.keys():\n",
    "        mod_obj = i.modified_bases[key]\n",
    "        mod_info = [m for m in mod_obj if m[0] == mod_pos[0][0]]\n",
    "        if len(mod_info) > 0:\n",
    "            modprob = (mod_info[0][1]+1)/256\n",
    "    return(modprob)\n",
    " \n",
    "def extract_dorado_calls_mer(bamFile, sample, threshold = 0.98, key = ('T', 0, 17802)):\n",
    "    samfile = pysam.AlignmentFile(bamFile, \"rb\")\n",
    "    tmp_EGFP_out = pd.DataFrame()\n",
    "    tmp_mCherry_out = pd.DataFrame()\n",
    "    thresholds_hist1 = []\n",
    "    thresholds_hist2 = []\n",
    "\n",
    "    for pileupcolumn in samfile.pileup(\"EGFP\", start = 114, end = 115, truncate = True, max_depth = 10000000, min_base_quality=13): #defaul min_base_qualiy = 13\n",
    "        print(pileupcolumn.pos)\n",
    "        out_df_mod = {\"mismatch_C\" : 0, \"mismatch_AUG\" : 0, \"canonical\" : 0, \"deletion\" : 0, \"modified\" : 0, \"fail\"  : 0}\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            mod_pos = [m for m in pileupread.alignment.get_aligned_pairs(with_seq=True) if m[1] == pileupcolumn.pos]\n",
    "            i = pileupread.alignment\n",
    "            if i.is_unmapped or i.is_supplementary or i.is_secondary:\n",
    "                continue\n",
    "            if len(mod_pos) > 0:\n",
    "                out_df_mod = extract_modifcation_infos(i, out_df_mod, mod_pos, threshold, key)\n",
    "                thresholds_hist1.append(get_mod_probs(i, mod_pos, key))\n",
    "\n",
    "        tmp = pd.DataFrame.from_dict(out_df_mod, orient = \"index\")\n",
    "        tmp[\"ref_pos\"] = pileupcolumn.pos\n",
    "        tmp_EGFP_out = pd.concat([tmp, tmp_EGFP_out], axis = 0)\n",
    "    tmp_EGFP_out[\"Ref\"] = \"EGFP\"\n",
    "    thresholds_hist_df1 = pd.DataFrame({\"Ref\" :\"EGFP\", \"Contig\" : sample, \"ref_pos\" : 114, \"thresholds\" : thresholds_hist1})\n",
    "    if (key[0] == \"C\"):\n",
    "        out_name = \"thresholds_EGFP_\" + sample + \"_Ccalls.csv\"\n",
    "    else:\n",
    "        out_name = \"thresholds_EGFP_\" + sample + \"_Ucalls.csv\"\n",
    "\n",
    "    thresholds_hist_df1.to_csv(out_name, index =False)\n",
    "    \n",
    "    \n",
    "    for pileupcolumn in samfile.pileup(\"mCherry\", start = 564, end = 565, truncate = True, max_depth = 10000000, min_base_quality=13): #defaul min_base_qualiy = 13\n",
    "        print(pileupcolumn.pos)\n",
    "        out_df_mod = {\"mismatch_C\" : 0, \"mismatch_AUG\" : 0, \"canonical\" : 0, \"deletion\" : 0, \"modified\" : 0, \"fail\"  : 0}\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            mod_pos = [m for m in pileupread.alignment.get_aligned_pairs(with_seq=True) if m[1] == pileupcolumn.pos]\n",
    "            i = pileupread.alignment\n",
    "            if i.is_unmapped or i.is_supplementary or i.is_secondary:\n",
    "                continue\n",
    "            if len(mod_pos) > 0:\n",
    "                out_df_mod = extract_modifcation_infos(i, out_df_mod, mod_pos, threshold, key)\n",
    "                thresholds_hist2.append(get_mod_probs(i, mod_pos, key))\n",
    "\n",
    "        tmp = pd.DataFrame.from_dict(out_df_mod, orient = \"index\")\n",
    "        tmp[\"ref_pos\"] = pileupcolumn.pos\n",
    "        tmp_mCherry_out = pd.concat([tmp, tmp_mCherry_out], axis = 0)\n",
    "    tmp_mCherry_out[\"Ref\"] = \"mCherry\"\n",
    "    thresholds_hist_df2 = pd.DataFrame({\"Ref\" :\"mCherry\", \"Contig\" : sample, \"ref_pos\" : 564, \"thresholds\" : thresholds_hist2})\n",
    "    if (key[0] == \"C\"):\n",
    "        out_name = \"thresholds_mCherry_\" + sample + \"_Ccalls.csv\"\n",
    "    else: \n",
    "        out_name = \"thresholds_mCherry_\" + sample + \"_Ucalls.csv\"\n",
    "    thresholds_hist_df2.to_csv(out_name, index =False)\n",
    "  \n",
    "    \n",
    "    out_df = pd.concat([tmp_mCherry_out.reset_index(), tmp_EGFP_out.reset_index()], axis = 0)\n",
    "    out_df[\"Contig\"] = sample\n",
    "    out_df.columns = [\"condition\", \"Nreads\", \"ref_pos\", \"Ref\", \"Contig\"]\n",
    "    return(out_df)\n",
    "\n",
    "def extract_dorado_calls_oligos_10mer(bamFile, sample, refs, threshold = 0.98, key = ('T', 0, 17802)):\n",
    "    samfile = pysam.AlignmentFile(bamFile, \"rb\")\n",
    "    tmp_1_out = pd.DataFrame()\n",
    "    tmp_2_out = pd.DataFrame()\n",
    "    thresholds_hist1 = []\n",
    "    thresholds_hist2 = []\n",
    "\n",
    "    for pileupcolumn in samfile.pileup(sample, start = 73, end = 74, truncate = True, max_depth = 10000000, min_base_quality=13): #defaul min_base_qualiy = 13\n",
    "        print(pileupcolumn.pos)\n",
    "        out_df_mod = {\"mismatch_C\" : 0, \"mismatch_AUG\" : 0, \"canonical\" : 0, \"deletion\" : 0, \"modified\" : 0, \"fail\"  : 0}\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            \n",
    "            i = pileupread.alignment\n",
    "            if i.is_unmapped or i.is_supplementary or i.is_secondary:\n",
    "                continue\n",
    "            mod_pos = [m for m in pileupread.alignment.get_aligned_pairs(with_seq=True) if m[1] == pileupcolumn.pos]\n",
    "            if len(mod_pos) > 0:\n",
    "                out_df_mod = extract_modifcation_infos(i, out_df_mod, mod_pos, threshold, key)\n",
    "                thresholds_hist1.append(get_mod_probs(i, mod_pos, key))\n",
    "\n",
    "        tmp = pd.DataFrame.from_dict(out_df_mod, orient = \"index\")\n",
    "        tmp[\"ref_pos\"] = pileupcolumn.pos\n",
    "        tmp_1_out = pd.concat([tmp, tmp_1_out], axis = 0)\n",
    "    tmp_1_out[\"Ref\"] = refs[0]\n",
    "    thresholds_hist_df1 = pd.DataFrame({\"Ref\" : refs[0], \"Contig\" : sample, \"ref_pos\" : 73, \"thresholds\" : thresholds_hist1})\n",
    "    if (key[0] == \"C\"):\n",
    "\n",
    "        out_name = \"thresholds_\" + refs[0] + sample + \"_Ccalls.csv\"\n",
    "    else:\n",
    "        out_name = \"thresholds_\" + refs[0] + sample + \"_Ucalls.csv\"\n",
    "\n",
    "    thresholds_hist_df1.to_csv(out_name, index =False)\n",
    "    \n",
    "    for pileupcolumn in samfile.pileup(sample, start = 113, end = 114, truncate = True, max_depth = 10000000, min_base_quality=13): #defaul min_base_qualiy = 13\n",
    "        print(pileupcolumn.pos)\n",
    "        out_df_mod = {\"mismatch_C\" : 0, \"mismatch_AUG\" : 0, \"canonical\" : 0, \"deletion\" : 0, \"modified\" : 0, \"fail\"  : 0}\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            mod_pos = [m for m in pileupread.alignment.get_aligned_pairs(with_seq=True) if m[1] == pileupcolumn.pos]\n",
    "            i = pileupread.alignment\n",
    "            if i.is_unmapped or i.is_supplementary or i.is_secondary:\n",
    "                continue\n",
    "            if len(mod_pos) > 0:\n",
    "                out_df_mod = extract_modifcation_infos(i, out_df_mod, mod_pos, threshold, key)\n",
    "                thresholds_hist2.append(get_mod_probs(i, mod_pos, key))\n",
    "\n",
    "        tmp = pd.DataFrame.from_dict(out_df_mod, orient = \"index\")\n",
    "        tmp[\"ref_pos\"] = pileupcolumn.pos\n",
    "        tmp_2_out = pd.concat([tmp, tmp_2_out], axis = 0)\n",
    "    tmp_2_out[\"Ref\"] = refs[1]\n",
    "    thresholds_hist_df2 = pd.DataFrame({\"Ref\" : refs[1], \"Contig\" : sample, \"ref_pos\" : 113, \"thresholds\" : thresholds_hist2})\n",
    "    if (key[0] == \"C\"):\n",
    "\n",
    "        out_name2 = \"thresholds_\" + refs[1] + sample + \"_Ccalls.csv\"\n",
    "    else: \n",
    "        out_name2 = \"thresholds_\" + refs[1] + sample + \"_Ucalls.csv\"\n",
    "    thresholds_hist_df2.to_csv(out_name2, index =False)\n",
    "    \n",
    "    \n",
    "    out_df = pd.concat([tmp_1_out.reset_index(), tmp_2_out.reset_index()], axis = 0)\n",
    "    out_df[\"Contig\"] = sample\n",
    "    out_df.columns = [\"condition\", \"Nreads\", \"ref_pos\", \"Ref\", \"Contig\"]\n",
    "    \n",
    "    \n",
    "    return(out_df)\n",
    "\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "012cb583",
   "metadata": {},
   "source": [
    "### 1. HEK293T Samples A and B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c54ab1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dorado basecalled samples using default settings for pseU calling\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/A_vector_004.0.7.2.vector.bam\"\n",
    "sample = \"Sample A\"\n",
    "Sample_A_mpileup = extract_dorado_calls_mer(bamFile, sample)\n",
    "\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/B_vector_004.0.7.2.vector.bam\"\n",
    "sample = \"Sample B\"\n",
    "Sample_B_mpileup = extract_dorado_calls_mer(bamFile, sample)\n",
    "\n",
    "samples_10mer = pd.concat([Sample_A_mpileup, Sample_B_mpileup], axis = 0)\n",
    "\n",
    "\n",
    "# Dorado basecalled samples using custom settings to call pseU on C sites (modification ID = 17803)\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/Y_eval.A_vector_004.0.7.2.vector.bam\"\n",
    "sample = \"Sample A\"\n",
    "Sample_A_mpileup2 = extract_dorado_calls_mer(bamFile, sample, key = ('C', 0, 17803))\n",
    "\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/Y_eval.B_vector_004.0.7.2.vector.bam\"\n",
    "sample = \"Sample B\"\n",
    "Sample_B_mpileup2 = extract_dorado_calls_mer(bamFile, sample, key = ('C', 0, 17803))\n",
    "\n",
    "samples_10mer2 = pd.concat([Sample_A_mpileup2, Sample_B_mpileup2], axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c368cce0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Removing T to A or G mismatches and deletions to calculate frequency of modification only on high quality canonical bases and modified bases\n",
    "samples_mpileup_flat = samples_10mer[~(samples_10mer[\"condition\"].isin([\"deletion\", \"mismatch_AUG\"]))]\n",
    "samples_mpileup_flat2 = samples_10mer2[~(samples_10mer2[\"condition\"].isin([\"deletion\", \"mismatch_AUG\"]))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c309ba5a",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.merge(samples_mpileup_flat, samples_mpileup_flat2,   on=['condition', 'ref_pos', 'Ref', 'Contig'], suffixes=[\"U_calls\", \"C_calls\"])\n",
    "df.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_TABLE_pseU_calls_samples.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f1269773",
   "metadata": {},
   "source": [
    "### 2. Synthetic Oligos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d37e620f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dorado basecalled samples using default settings for pseU calling\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/Oligo1.0.7.2.oligo.bam\"\n",
    "sample = \"F1_GFPPsi_mCherry\"\n",
    "refs = [\"EGFP\", \"mCherry\"]\n",
    "GFP_mpileup = extract_dorado_calls_oligos_10mer(bamFile, sample,refs)\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/Oligo2.0.7.2.oligo.bam\"\n",
    "sample = \"F3_mCherryPsi_GFP\"\n",
    "refs = [\"mCherry\", \"EGFP\"]\n",
    "mCherry_mpileup = extract_dorado_calls_oligos_10mer(bamFile, sample, refs)\n",
    "\n",
    "oligos_10mer = pd.concat([mCherry_mpileup, GFP_mpileup], axis = 0)\n",
    "\n",
    "# Dorado basecalled samples using custom settings to call pseU on C sites (modification ID = 17803)\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/Y_eval.Oligo1.0.7.2.oligo.bam\"\n",
    "sample = \"F1_GFPPsi_mCherry\"\n",
    "refs = [\"EGFP\", \"mCherry\"]\n",
    "GFP_mpileup2 = extract_dorado_calls_oligos_10mer(bamFile, sample,refs, key = ('C', 0, 17803))\n",
    "bamFile = \"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/LEMKE_oligos_072_felix/Y_eval.Oligo2.0.7.2.oligo.bam\"\n",
    "sample = \"F3_mCherryPsi_GFP\"\n",
    "refs = [\"mCherry\", \"EGFP\"]\n",
    "mCherry_mpileup2 = extract_dorado_calls_oligos_10mer(bamFile, sample, refs, key = ('C', 0, 17803))\n",
    "\n",
    "oligos_10mer2 = pd.concat([mCherry_mpileup2, GFP_mpileup2], axis = 0)\n",
    "\n",
    "# Removing T to A or G mismatches and deletions to calculate frequency of modification only on high quality canonical bases and modified bases\n",
    "oligos_mpileup_flat = oligos_10mer[~(oligos_10mer[\"condition\"].isin([\"deletion\", \"mismatch_AUG\"]))]\n",
    "oligos_mpileup_flat2 = oligos_10mer2[~(oligos_10mer2[\"condition\"].isin([\"deletion\", \"mismatch_AUG\"]))]\n",
    "\n",
    "df = pd.merge(oligos_mpileup_flat, oligos_mpileup_flat2,   on=['condition', 'ref_pos', 'Ref', 'Contig'], suffixes=[\"U_calls\", \"C_calls\"])\n",
    "df.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_pseU_calls_oligos.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79d27fd5",
   "metadata": {},
   "source": [
    "### 3. Gene Regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d45e3c63",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mod_probs(i, mod_pos, key = ('T', 0, 17802)):\n",
    "    modprob = None\n",
    "    if key in i.modified_bases.keys():\n",
    "        mod_obj = i.modified_bases[key]\n",
    "        mod_info = [m for m in mod_obj if m[0] == mod_pos[0][0]]\n",
    "        if len(mod_info) > 0:\n",
    "            modprob = (mod_info[0][1]+1)/256\n",
    "    return(modprob)\n",
    "def extract_modifcation_infos_reverse(i, out_df_mod, mod_pos, threshold, key = ('T', 1, 17802)):\n",
    "    #print(i.modified_bases.keys())\n",
    "    if key not in i.modified_bases.keys():\n",
    "\n",
    "        return(out_df_mod)\n",
    "    mod_obj = i.modified_bases[key]\n",
    "    mod_info = [m for m in mod_obj if m[0] == mod_pos[0][0]]\n",
    "    if len(mod_info) == 0:\n",
    "        mod_info = [(mod_pos[0][0],0)]\n",
    "        modprob = 0\n",
    "        #print(mod_pos)\n",
    "        if mod_pos[0][2] in [\"t\", \"a\", \"g\", \"c\"]:\n",
    "            called_base = i.query_sequence[mod_pos[0][0]]\n",
    "            if called_base == \"G\":\n",
    "                out_df_mod[\"mismatch_C\"] += 1\n",
    "            else:\n",
    "                out_df_mod[\"mismatch_AUG\"] += 1\n",
    "        elif mod_pos[0][0] == None:\n",
    "            out_df_mod[\"deletion\"] += 1\n",
    "        else:\n",
    "            out_df_mod[\"canonical\"] += 1\n",
    "   \n",
    "    else:\n",
    "        modprob = (mod_info[0][1]+1)/256\n",
    "        if modprob < threshold:\n",
    "            out_df_mod[\"fail\"] += 1\n",
    "        else:\n",
    "            out_df_mod[\"modified\"] += 1\n",
    "    #print(out_df_mod)\n",
    "    #print(len(out_df_mod))\n",
    "    return(out_df_mod)\n",
    "def extract_dorado_calls_10mer_single_region(bamFile, sample,positions, threshold = 0.98, key = ('T', 1, 17802)):\n",
    "    samfile = pysam.AlignmentFile(bamFile, \"rb\")\n",
    "    out_df = pd.DataFrame()\n",
    "    thresholds_hist = []\n",
    "    print(positions)\n",
    "    #return(samfile)\n",
    "    for pileupcolumn in samfile.pileup(positions[0], start = positions[1], end = positions[2], truncate = True, max_depth = 10000000, min_base_quality=13): #defaul min_base_qualiy = 13\n",
    "        print(pileupcolumn.pos)\n",
    "        out_df_mod = {\"mismatch_C\" : 0, \"mismatch_AUG\" : 0, \"canonical\" : 0, \"deletion\" : 0, \"modified\" : 0, \"fail\"  : 0}\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            mod_pos = [m for m in pileupread.alignment.get_aligned_pairs(with_seq=True) if m[1] == pileupcolumn.pos]\n",
    "            i = pileupread.alignment\n",
    "            if i.is_unmapped or i.is_supplementary or i.is_secondary:\n",
    "                continue\n",
    "            if len(mod_pos) > 0:\n",
    "                if mod_pos[0][0] == None:\n",
    "                    continue\n",
    "                out_df_mod = extract_modifcation_infos_reverse(i, out_df_mod, mod_pos, threshold, key)\n",
    "                thresholds_hist.append(get_mod_probs(i, mod_pos, key))\n",
    "\n",
    "        tmp = pd.DataFrame.from_dict(out_df_mod, orient = \"index\")\n",
    "        #print(tmp)\n",
    "        tmp[\"ref_pos\"] = pileupcolumn.pos\n",
    "        out_df = pd.concat([tmp, out_df], axis = 0)\n",
    "    thresholds_hist_df = pd.DataFrame({\"Ref\" :\"PSMB2\", \"Contig\" : sample, \"ref_pos\" : pileupcolumn.pos, \"thresholds\" : thresholds_hist})\n",
    "    if (key[0] == \"C\"):\n",
    "        out_name = \"thresholds_PSMB2_\" + sample + \"_Ccalls.csv\"\n",
    "    else:\n",
    "        out_name = \"thresholds_PSMB2_\" + sample + \"_Ucalls.csv\"\n",
    "\n",
    "    thresholds_hist_df.to_csv(out_name, index =False)\n",
    "    \n",
    "    out_df = out_df.reset_index()\n",
    "    out_df[\"Ref\"] = sample\n",
    "    out_df[\"Contig\"] = sample\n",
    "    out_df.columns = [\"condition\", \"Nreads\", \"ref_pos\", \"Ref\", \"Contig\"]\n",
    "    return(out_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "781ca227",
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "import os.path\n",
    "files = glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/DORADO_0_7_2_folder/DUSP5_folder/RNA004*.new.PSMB2.bam\")\n",
    "\n",
    "out_PSMB2 = pd.DataFrame()\n",
    "for bamFile in files:\n",
    "    baseName = os.path.basename(bamFile)\n",
    "    baseName = baseName.replace(\"_basecall\", \"\")\n",
    "    sample = baseName.replace(\".0.7.2.GRCh38.new.PSMB2.bam\", \"\")\n",
    "    print(sample)\n",
    "    subsample_mpileup = extract_dorado_calls_10mer_single_region(bamFile, sample, positions=[\"chr1\", 35603332, 35603333])\n",
    "    out_PSMB2 = pd.concat([out_PSMB2, subsample_mpileup], axis = 0)\n",
    "out_PSMB2[\"Gene\"] = \"PSMB2\"\n",
    "\n",
    "\n",
    "files = glob.glob(\"/raid/chhewel_analysis/RNA004_Manuscript_reanalysis/DORADO_0_7_2_folder/DUSP5_folder/Y*RNA004*.aligned.bam\")\n",
    "\n",
    "out_PSMB2_2 = pd.DataFrame()\n",
    "for bamFile in files:\n",
    "    baseName = os.path.basename(bamFile)\n",
    "    baseName = baseName.replace(\"_basecall\", \"\")\n",
    "    baseName = baseName.replace(\"Y_eval.\", \"\")\n",
    "\n",
    "    sample = baseName.replace(\".0.7.2.GRCh38.pod5.aligned.bam\", \"\")\n",
    "    print(sample)\n",
    "    subsample_mpileup = extract_dorado_calls_10mer_single_region(bamFile, sample, positions=[\"chr1\", 35603332, 35603333], key = ('C', 1, 17803))\n",
    "    out_PSMB2_2 = pd.concat([out_PSMB2_2, subsample_mpileup], axis = 0)\n",
    "out_PSMB2_2[\"Gene\"] = \"PSMB2\"\n",
    "\n",
    "\n",
    "PSMB2_mpileup_flat = out_PSMB2[~(out_PSMB2[\"condition\"].isin([\"deletion\", \"mismatch_AUG\"]))]\n",
    "PSMB2_mpileup_flat2 = out_PSMB2_2[~(out_PSMB2_2[\"condition\"].isin([\"deletion\", \"mismatch_AUG\"]))]\n",
    "\n",
    "\n",
    "df = pd.merge(PSMB2_mpileup_flat, PSMB2_mpileup_flat2,   on=['condition', 'ref_pos', 'Ref', 'Contig', 'Gene'], suffixes=[\"U_calls\", \"C_calls\"])\n",
    "df.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_pseU_calls_PSMB2.tsv\", sep = \"\\t\", index = False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "544f88e5",
   "metadata": {},
   "source": [
    "## 2. MODIFICATION PROBABILITY PLOTS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "060d09c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Modification probability tables are generated using pysam (see above)\n",
    "\n",
    "#__________________________________________________________________\n",
    "## SYNTHETIC OLGIOS\n",
    "# Load probability files generated by pysam (see above) \n",
    "mod_EGFP = pd.read_csv(\"thresholds_EGFPF1_GFPPsi_mCherry_Ucalls.csv\")\n",
    "unmod_EGFP = pd.read_csv(\"thresholds_EGFPF3_mCherryPsi_GFP_Ucalls.csv\")\n",
    "unmod_mCherry = pd.read_csv(\"thresholds_mCherryF1_GFPPsi_mCherry_Ucalls.csv\")\n",
    "mod_mCherry = pd.read_csv(\"thresholds_mCherryF3_mCherryPsi_GFP_Ucalls.csv\")\n",
    "\n",
    "# Load probability (pseU calling on C-sites) files generated by pysam (see above) \n",
    "mod_EGFP_C = pd.read_csv(\"thresholds_EGFPF1_GFPPsi_mCherry_Ccalls.csv\")\n",
    "unmod_EGFP_C = pd.read_csv(\"thresholds_EGFPF3_mCherryPsi_GFP_Ccalls.csv\")\n",
    "unmod_mCherry_C = pd.read_csv(\"thresholds_mCherryF1_GFPPsi_mCherry_Ccalls.csv\")\n",
    "mod_mCherry_C = pd.read_csv(\"thresholds_mCherryF3_mCherryPsi_GFP_Ccalls.csv\")\n",
    "\n",
    "mod_EGFP_flat = mod_EGFP.dropna()\n",
    "unmod_EGFP_flat = unmod_EGFP.dropna()\n",
    "unmod_mCherry_flat = unmod_mCherry.dropna()\n",
    "mod_mCherry_flat = mod_mCherry.dropna()\n",
    "\n",
    "mod_EGFP_flat[\"Sample\"] = \"Oligo1_Pos74\"\n",
    "mod_mCherry_flat[\"Sample\"] = \"Oligo2_Pos74\"\n",
    "mod_distr = pd.concat([mod_EGFP_flat, mod_mCherry_flat], axis = 0)\n",
    "\n",
    "mod_EGFP_C_flat = mod_EGFP_C.dropna()\n",
    "unmod_EGFP_C_flat = unmod_EGFP_C.dropna()\n",
    "unmod_mCherry_C_flat = unmod_mCherry_C.dropna()\n",
    "mod_mCherry_C_flat = mod_mCherry_C.dropna()\n",
    "\n",
    "mod_EGFP_C_flat[\"Sample\"] = \"Oligo1_Pos74\"\n",
    "mod_mCherry_C_flat[\"Sample\"] = \"Oligo2_Pos74\"\n",
    "mod_distr_C = pd.concat([mod_EGFP_C_flat, mod_mCherry_C_flat], axis = 0)\n",
    "\n",
    "# OLIGOS POS 74 Ucalls\n",
    "plt.figure(figsize=(5, 4))  # 10 inches wide by 6 inches tall\n",
    "sns.histplot(data=mod_distr, x=\"thresholds\", hue = \"Sample\", stat = \"density\")\n",
    "\n",
    "plt.savefig(\"PAPER_thresholds_oligos_pos74_Ucalls.svg\",bbox_inches='tight', dpi=300)\n",
    "plt.savefig(\"PAPER_thresholds_oligos_pos74_Ucalls.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "# OLIGOS POS 74 Ccalls\n",
    "plt.figure(figsize=(5, 4))  # 10 inches wide by 6 inches tall\n",
    "sns.histplot(data=mod_distr_C, x=\"thresholds\", hue = \"Sample\", stat = \"density\")\n",
    "\n",
    "plt.savefig(\"PAPER_thresholds_oligos_pos74_Ccalls.svg\",bbox_inches='tight', dpi=300)\n",
    "plt.savefig(\"PAPER_thresholds_oligos_pos74_Ccalls.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "#__________________________________________________________________\n",
    "## REPORTERS: EGFP AND MCHERRY\n",
    "# Load files \n",
    "EGFP_SampleA = pd.read_csv(\"thresholds_EGFP_Sample A_Ucalls.csv\")\n",
    "EGFP_SampleB = pd.read_csv(\"thresholds_EGFP_Sample B_Ucalls.csv\")\n",
    "mCherry_SampleA = pd.read_csv(\"thresholds_mCherry_Sample A_Ucalls.csv\")\n",
    "mCherry_SampleB = pd.read_csv(\"thresholds_mCherry_Sample B_Ucalls.csv\")\n",
    "\n",
    "# Load files \n",
    "EGFP_SampleA_C = pd.read_csv(\"thresholds_EGFP_Sample A_Ccalls.csv\")\n",
    "EGFP_SampleB_C = pd.read_csv(\"thresholds_EGFP_Sample B_Ccalls.csv\")\n",
    "mCherry_SampleA_C = pd.read_csv(\"thresholds_mCherry_Sample A_Ccalls.csv\")\n",
    "mCherry_SampleB_C = pd.read_csv(\"thresholds_mCherry_Sample B_Ccalls.csv\")\n",
    "\n",
    "EGFP_SampleA_flat = EGFP_SampleA.dropna()\n",
    "EGFP_SampleB_flat = EGFP_SampleB.dropna()\n",
    "mCherry_SampleA_flat = mCherry_SampleA.dropna()\n",
    "mCherry_SampleB_flat = mCherry_SampleB.dropna()\n",
    "\n",
    "EGFP_SampleA_flat[\"Sample\"] = \"SampleA_EGFP\"\n",
    "mCherry_SampleA_flat[\"Sample\"] = \"SampleA_mCherry\"\n",
    "EGFP_SampleB_flat[\"Sample\"] = \"SampleB_EGFP\"\n",
    "mCherry_SampleB_flat[\"Sample\"] = \"SampleB_mCherry\"\n",
    "Samples_thresholds = pd.concat([EGFP_SampleA_flat, mCherry_SampleA_flat, EGFP_SampleB_flat, mCherry_SampleB_flat], axis = 0)\n",
    "\n",
    "EGFP_SampleA_C_flat = EGFP_SampleA_C.dropna()\n",
    "EGFP_SampleB_C_flat = EGFP_SampleB_C.dropna()\n",
    "mCherry_SampleA_C_flat = mCherry_SampleA_C.dropna()\n",
    "mCherry_SampleB_C_flat = mCherry_SampleB_C.dropna()\n",
    "\n",
    "EGFP_SampleA_C_flat[\"Sample\"] = \"SampleA_EGFP\"\n",
    "mCherry_SampleA_C_flat[\"Sample\"] = \"SampleA_mCherry\"\n",
    "EGFP_SampleB_C_flat[\"Sample\"] = \"SampleB_EGFP\"\n",
    "mCherry_SampleB_C_flat[\"Sample\"] = \"SampleB_mCherry\"\n",
    "Samples_thresholds_C = pd.concat([EGFP_SampleA_C_flat, mCherry_SampleA_C_flat, EGFP_SampleB_C_flat, mCherry_SampleB_C_flat], axis = 0)\n",
    "\n",
    "# U calls\n",
    "plt.figure(figsize=(5, 4))  # 10 inches wide by 6 inches tall\n",
    "sns.histplot(data=Samples_thresholds, x=\"thresholds\", hue = \"Sample\", stat = \"density\")\n",
    "\n",
    "plt.savefig(\"PAPER_thresholds_samples_Ucalls.svg\",bbox_inches='tight', dpi=300)\n",
    "plt.savefig(\"PAPER_thresholds_samples_Ucalls.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "# C calls\n",
    "plt.figure(figsize=(5, 4))  # 10 inches wide by 6 inches tall\n",
    "sns.histplot(data=Samples_thresholds_C, x=\"thresholds\", hue = \"Sample\", stat = \"density\")\n",
    "\n",
    "plt.savefig(\"PAPER_thresholds_samples_Ccalls.svg\",bbox_inches='tight', dpi=300)\n",
    "plt.savefig(\"PAPER_thresholds_samples_Ccalls.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "#__________________________________________________________________\n",
    "## PSMB2 SITE \n",
    "TS_out_C = pd.DataFrame()\n",
    "TS_out_U = pd.DataFrame()\n",
    "files = glob.glob(\"/raid/awiercze_analysis/RNA004_manuscript_scripts/FINAL_SCRIPTS/thresholds_PSMB2_*calls.csv\")\n",
    "b = \"RNA004_E\"\n",
    "c = \"ribo\"\n",
    "files = [i for i in files if b not in i and c not in i]\n",
    "# Load files\n",
    "for treshfile in files:\n",
    "    m = pd.read_csv(treshfile)\n",
    "    m_sub = m.dropna()\n",
    "    if m_sub.empty:\n",
    "        continue\n",
    "    m_sub['Sample'] = treshfile.replace(\"/raid/awiercze_analysis/RNA004_manuscript_scripts/FINAL_SCRIPTS/\", \"\").replace(\"thresholds_PSMB2_\", \"\").replace(\".csv\", \"\")\n",
    "    print(m_sub['Sample'].unique())\n",
    "    if \"Ucalls\" in m_sub['Sample'].unique()[0]:\n",
    "        print(\"Ucalls\")\n",
    "        TS_out_U = pd.concat([TS_out_U, m_sub], axis = 0)\n",
    "    else:\n",
    "        print(\"Ccalls\")\n",
    "\n",
    "        TS_out_C = pd.concat([TS_out_C, m_sub], axis = 0)\n",
    "\n",
    "# U_calls\n",
    "plt.figure(figsize=(5, 4))  # 10 inches wide by 6 inches tall\n",
    "sns.histplot(data=TS_out_U, x=\"thresholds\", hue = \"Sample\", stat = \"density\")\n",
    "\n",
    "plt.savefig(\"PAPER_thresholds_PSMB2_Ucalls.svg\",bbox_inches='tight', dpi=300)\n",
    "plt.savefig(\"PAPER_thresholds_PSMB2_Ucalls.pdf\", dpi=300, bbox_inches='tight')\n",
    "\n",
    "# C_calls\n",
    "plt.figure(figsize=(5, 4))  # 10 inches wide by 6 inches tall\n",
    "sns.histplot(data=TS_out_C, x=\"thresholds\", hue = \"Sample\", stat = \"density\")\n",
    "\n",
    "plt.savefig(\"PAPER_thresholds_PSMB2_Ccalls.svg\",bbox_inches='tight', dpi=300)\n",
    "plt.savefig(\"PAPER_thresholds_PSMB2_Ccalls.pdf\", dpi=300, bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ed8ef95",
   "metadata": {},
   "source": [
    "## 3. m6A modification frequency at Position A1832 on 18S rRNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fea4547b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#_______________________________________________________________\n",
    "# M6A MODIFICATION FREQUENCY ON A1832 OF 18S RRNA                         #\n",
    "\n",
    "# LOAD LIBRARIES\n",
    "import pandas as pd\n",
    "import pysam\n",
    "import seaborn as sns\n",
    "import matplotlib.pylab as plt\n",
    "from scipy.stats import ranksums\n",
    "import numpy as np\n",
    "\n",
    "# INPUT BAM FILES BASECALLED BY DORADO INCLUDING M6A CALLING\n",
    "JN1328_24=\"filter.1328_24_mettl5.072.rDNA.bam\"\n",
    "patient=\"filter.Patient_mettl5.072.rDNA.bam\"\n",
    "control2=\"filter.PB_mettl5.072.rDNA.bam\"\n",
    "IVT=\"IVT_18S.filteredrDNA.bam\"\n",
    "HEK_C=\"RNA004_HEK293T_C.rDNA.filter.bam\"\n",
    "HEK_A=\"RNA004_HEK293T_A.rDNA.filter.bam\"\n",
    "HEK_B=\"RNA004_HEK293T_B.rDNA.filter.bam\"\n",
    "#UHRR=\"RNA004_UHRR_3.rDNA.filter.bam\"\n",
    "\n",
    "\n",
    "# FUNCTION TO EXTRACT m6A MODIFICATION FREQUENCY\n",
    "def extract_mod_probs_m6A(bam_path, sample, key = ('A', 0, 'a'), chrom = \"chrR\", start = 14826, end = 14827):\n",
    "    d = []\n",
    "    samfile = pysam.AlignmentFile(bam_path, \"rb\")\n",
    "    for pileupcolumn in samfile.pileup(chrom, start = start, end = end, truncate = True, max_depth = 3000000, min_base_quality=13): #defaul min_base_qualiy = 13\n",
    "        print(pileupcolumn.pos)\n",
    "        print(pileupcolumn.n)\n",
    "        n = 0\n",
    "        for pileupread in pileupcolumn.pileups:\n",
    "            n +=1\n",
    "            mod_pos = [m for m in pileupread.alignment.get_aligned_pairs(with_seq=True) if m[1] == pileupcolumn.pos]\n",
    "            i = pileupread.alignment\n",
    "            if i.is_unmapped or i.is_supplementary or i.is_secondary:\n",
    "                continue\n",
    "            if len(mod_pos) > 0:\n",
    "                #print(i)\n",
    "                if (key not in i.modified_bases.keys()):\n",
    "                    continue\n",
    "                mod_obj = i.modified_bases[key]\n",
    "                #print(mod_pos)\n",
    "                mod_info = [m for m in mod_obj if m[0] == mod_pos[0][0]]\n",
    "                #print(mod_info)\n",
    "                #print(mod_pos[0][2])\n",
    "                if len(mod_info) == 0:\n",
    "                    mod_info = [(mod_pos[0][0],0)]\n",
    "                    modprob = 0\n",
    "                    #print(mod_pos)\n",
    "                    if mod_pos[0][2] in [\"t\", \"a\", \"g\", \"c\"]:\n",
    "                        called_base = i.query_sequence[mod_pos[0][0]]\n",
    "                        \n",
    "                    elif mod_pos[0][0] == None:\n",
    "                        called_base = \"-\"\n",
    "                    else:\n",
    "                        called_base = mod_pos[0][2]\n",
    "            \n",
    "                else:\n",
    "                    modprob = (mod_info[0][1]+1)/256\n",
    "                    called_base = mod_pos[0][2]\n",
    "                d.append({\"Sample\": sample, \"Position\" : pileupcolumn.pos, \"Mod_prob\" : modprob, \"Called_base\" : called_base})\n",
    "    d_df = pd.DataFrame(d)\n",
    "    samfile.close()\n",
    "    print(n)\n",
    "    return(d_df)\n",
    "\n",
    "\n",
    "# EXTRACT MODIFICATION FREQUENCY\n",
    "filepath = \"/home/chhewel/mettl5_update/\"\n",
    "d_df_patient = extract_mod_probs_m6A(filepath + patient, \"Patient\")\n",
    "d_df_JN1328_24 = extract_mod_probs_m6A(filepath + JN1328_24, \"Healthy child\")\n",
    "d_df_Control2 = extract_mod_probs_m6A(filepath + control2, \"Healthy adult\")\n",
    "d_df_IVT = extract_mod_probs_m6A(filepath + \"last_aligns/\" + IVT, \"IVT\", chrom = \"RNA45SN1\", start = 5485, end = 5486)\n",
    "#d_df_HEKC_C = extract_mod_probs_m6A(\"/home/awiercze/\" + HEK_C, \"HEK293T_C\")\n",
    "\n",
    "\n",
    "d_df_patient_match = d_df_patient[d_df_patient[\"Called_base\"] == \"A\"]\n",
    "d_df_JN1328_24_match = d_df_JN1328_24[d_df_JN1328_24[\"Called_base\"] == \"A\"]\n",
    "d_df_Control2_match = d_df_Control2[d_df_Control2[\"Called_base\"] == \"A\"]\n",
    "d_df_IVT_match = d_df_IVT[d_df_IVT[\"Called_base\"] == \"A\"]\n",
    "\n",
    "\n",
    "d_out = pd.concat([d_df_patient_match, d_df_JN1328_24_match, d_df_Control2_match, d_df_IVT_match], axis = 0)\n",
    "\n",
    "# CALCULATE STATS \n",
    "stats_agecontrol, pvalue_agecontrol = ranksums(d_df_patient_match[\"Mod_prob\"], d_df_JN1328_24_match[\"Mod_prob\"], alternative = \"two-sided\")\n",
    "stats_control, pvalue_control = ranksums(d_df_patient_match[\"Mod_prob\"], d_df_Control2_match[\"Mod_prob\"], alternative = \"two-sided\")\n",
    "stats_IVT, pvalue_IVT = ranksums(d_df_patient_match[\"Mod_prob\"], d_df_IVT_match[\"Mod_prob\"], alternative = \"two-sided\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d39e2854",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PLOT RESULTS\n",
    "input_colors = {\"RNA002_blood\" :\"#cc78bc\", \"IVT\" :\"#029e73\"}\n",
    "input_colors = {\"IVT\" :\"#8B5CF6\", \"Patient\" :\"#EE6C4D\", \"Healthy child\" : \"#457b9dff\", \"Healthy adult\" : \"#254470ff\"}\n",
    "plt.figure(figsize=(4, 4))\n",
    "d_out[\"prob_perc\"] = d_out[\"Mod_prob\"]* 100\n",
    "ax = sns.violinplot(x='Sample', y='prob_perc', inner = \"quartile\", palette=input_colors, hue = \"Sample\", data=d_out, order = [\"Patient\", \"IVT\", \"Healthy child\", \"Healthy adult\"], cut = 0)\n",
    "\n",
    "ax.set_ylabel(\"m6A modification probability [%]\\n(Dorado)\")\n",
    "ax.set_xlabel(\"\")\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation = 45, ha = \"right\", rotation_mode = \"anchor\")\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "ax.set_ylim(0 , 100)\n",
    "plt.title('m6A probability at position A1832 on 18S rRNA')\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE6/PAPER_FIGURE_6E_prob.pdf\", dpi = 300, bbox_inches='tight', transparent = True)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE6/PAPER_FIGURE_6E_prob.png\", dpi = 300, bbox_inches='tight', transparent = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e43eb547",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_out['Sample'] = pd.Categorical(d_out['Sample'], categories=[\"Patient\", \"IVT\", \"Healthy child\", \"Healthy adult\", ], ordered=True)\n",
    "grouped = d_out.groupby('Sample')['Mod_prob']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a28e5086",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_percentage = grouped.apply(lambda x: (x >= 0.98).sum() / len(x) * 100).reset_index(name = \"m6A_freq\")  # Red\n",
    "num_reads = grouped.size().reset_index(name = \"num_reads\")  # \n",
    "new_df = pd.merge(match_percentage, num_reads)\n",
    "new_df.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_METTL5_freq.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ae96d94",
   "metadata": {},
   "outputs": [],
   "source": [
    "d_out.to_csv(\"~/NEW_RNA004_PLOTS/TABLES/PAPER_METTL5_raw.tsv\", sep = \"\\t\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57c88934",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_percentage = grouped.apply(lambda x: (x >= 0.98).sum() / len(x) * 100)  # Red\n",
    "zero_percentage = grouped.apply(lambda x: (x < 0.98).sum() / len(x) * 100)  # Green\n",
    "#non_match_percentage = 100 - match_percentage - zero_percentage  # Blue (remaining)\n",
    "\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(figsize=(4, 4))\n",
    "\n",
    "bars1 = ax.bar(match_percentage.index, match_percentage,edgecolor='black', color='#ee6c4d', label='m6A (p >= 0.98)')\n",
    "#bars2 = ax.bar(non_match_percentage.index, non_match_percentage,edgecolor='white', bottom=match_percentage, color='#0b3954', label='Failed (p < 0.98)')\n",
    "bars3 = ax.bar(zero_percentage.index, zero_percentage, edgecolor='black',\n",
    "       bottom=match_percentage,color='#a0a0a0', label='Not modified')\n",
    "\n",
    "\n",
    "for bars in [bars1, bars3]:  # Loop through all bars\n",
    "    for bar in bars:\n",
    "        height = bar.get_height()\n",
    "        if height > 0:  # Avoid labeling zero values\n",
    "              if height < 1:\n",
    "                     ax.text(bar.get_x() + bar.get_width()/2,  # X position (center of bar)\n",
    "                            bar.get_y() + 1.2,  # Y position (middle of bar)\n",
    "                            f'{height:.2f}%',  # Text (rounded percentage)\n",
    "                            ha='center', va='center', fontsize=8, color='black')\n",
    "              else:      \n",
    "                     ax.text(bar.get_x() + bar.get_width() / 2,  # X position (center of bar)\n",
    "                            bar.get_y() + height / 2,  # Y position (middle of bar)\n",
    "                            f'{height:.2f}%',  # Text (rounded percentage)\n",
    "                            ha='center', va='center', fontsize=8, color='black')\n",
    "\n",
    "\n",
    "\n",
    "# Labels & Formatting\n",
    "ax.set_ylabel('Percentage reads (%)', color = \"black\")\n",
    "ax.set_title('Percentage of modified bases at position A1832 on 18S rRNA', color = \"black\")\n",
    "ax.legend(loc = 1)\n",
    "ax.spines[\"top\"].set_visible(False)\n",
    "ax.spines[\"right\"].set_visible(False)\n",
    "plt.xticks(rotation=45, color = \"black\",ha='right')\n",
    "#plt.yticks(color = \"white\")\n",
    "\n",
    "fig.patch.set_alpha(0.0) \n",
    "plt.subplots_adjust(bottom=0.2)\n",
    "#plt.show()\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE6/PAPER_FIGURE_6E_new_black.png\", bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/FIGURE6/PAPER_FIGURE_6E_new_black.pdf\", bbox_inches='tight', transparent=True, dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad9e4350",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_percentage = grouped.apply(lambda x: (x >= 0.98).sum() / len(x) * 100)  # Red\n",
    "zero_percentage = grouped.apply(lambda x: (x == 0).sum() / len(x) * 100)  # Green\n",
    "non_match_percentage = 100 - match_percentage - zero_percentage  # Blue (remaining)\n",
    "\n",
    "\n",
    "# Plot\n",
    "fig, ax = plt.subplots(figsize=(3.5, 4.5))\n",
    "\n",
    "bars1 = ax.bar(match_percentage.index, match_percentage,edgecolor='black', color='#ee6c4d', label='m6A (p >= 0.98)')\n",
    "bars2 = ax.bar(non_match_percentage.index, non_match_percentage,edgecolor='black', bottom=match_percentage, color=\"#cab3ae\", label='Failed (p < 0.98)')\n",
    "bars3 = ax.bar(zero_percentage.index, zero_percentage, edgecolor='black',\n",
    "       bottom=match_percentage + non_match_percentage, color='#a0a0a0', label='Not modified')\n",
    "\n",
    "\n",
    "for bars in [bars1, bars2, bars3]:  # Loop through all bars\n",
    "    for bar in bars:\n",
    "        height = bar.get_height()\n",
    "        if height > 0:  # Avoid labeling zero values\n",
    "              if height < 1:\n",
    "                     ax.text(bar.get_x() + bar.get_width()/2,  # X position (center of bar)\n",
    "                            bar.get_y() + 1.2,  # Y position (middle of bar)\n",
    "                            f'{height:.2f}%',  # Text (rounded percentage)\n",
    "                            ha='center', va='center', fontsize=10, color='white')\n",
    "              else:      \n",
    "                     ax.text(bar.get_x() + bar.get_width() / 2,  # X position (center of bar)\n",
    "                            bar.get_y() + height / 2,  # Y position (middle of bar)\n",
    "                            f'{height:.2f}%',  # Text (rounded percentage)\n",
    "                            ha='center', va='center', fontsize=10, color='white')\n",
    "\n",
    "\n",
    "\n",
    "# Labels & Formatting\n",
    "ax.set_ylabel('Percentage reads (%)')\n",
    "#ax.set_title('Percentage of modified bases at position A1832 on 18S rRNA')\n",
    "plt.legend(\n",
    "    title=\"m6A cutoff\",\n",
    "    loc=\"center left\",\n",
    "    bbox_to_anchor=(0.15, -0.42),\n",
    "    borderaxespad=0\n",
    ")\n",
    "plt.xticks(ticks = [0,1,2,3], labels = [\"Patient\", \"IVT\", \"Healhty child\", \"Healthy adult\"], rotation=45, color = \"black\", ha='right')\n",
    "fig.patch.set_alpha(0.0) \n",
    "plt.subplots_adjust(bottom=0.2)\n",
    "#plt.show()\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/PAPER_FIGURE_6E_new_black.png\", bbox_inches='tight', transparent=True, dpi=300)\n",
    "plt.savefig(\"/home/awiercze/NEW_RNA004_PLOTS/PAPER_FIGURE_6E_new_black.pdf\", bbox_inches='tight', transparent=True, dpi=300)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
